// @bryntum/gantt-react/src/BryntumResourceUtilization.tsx
/**
 * React wrapper for Bryntum ResourceUtilization
 */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React, { RefObject } from 'react';
import { AjaxStore, AjaxStoreConfig, AssignmentAllocationInterval, Base, CellCopyPaste, CellCopyPasteConfig, CellMenu, CellMenuConfig, CellTooltip, CellTooltipConfig, Column, ColumnAutoWidth, ColumnAutoWidthConfig, ColumnDragToolbar, ColumnDragToolbarConfig, ColumnLines, ColumnLinesConfig, ColumnPicker, ColumnPickerConfig, ColumnRename, ColumnRenameConfig, ColumnReorder, ColumnReorderConfig, ColumnResize, ColumnResizeConfig, ColumnStore, ColumnStoreConfig, Container, ContainerItemConfig, CrudManager, DomConfig, DurationUnit, Editor, EventColor, ExportConfig, Field, FileDrop, FileDropConfig, FillHandle, FillHandleConfig, Filter, FilterBar, FilterBarConfig, FilterConfig, Grid, GridBase, GridColumnConfig, GridExcelExporter, GridExcelExporterConfig, GridLocation, GridLocationConfig, GridLockRows, GridLockRowsConfig, GridPdfExport, GridPdfExportConfig, GridPrint, GridPrintConfig, GridRowResize, GridRowResizeConfig, GridSelectionMode, GridTreeGroup, GridTreeGroupConfig, Group, GroupConfig, GroupSummary, GroupSummaryConfig, HeaderMenu, HeaderMenuConfig, HeaderZoom, HeaderZoomConfig, Histogram, HistogramConfig, HistogramSeries, KeyMapConfig, Mask, MaskConfig, Menu, MenuItem, MenuItemConfig, MergeCells, MergeCellsConfig, Model, ModelConfig, NonWorkingTime, NonWorkingTimeConfig, PagingToolbarConfig, Pan, PanConfig, Panel, PanelCollapserConfig, PanelCollapserOverlayConfig, PanelHeader, PreserveScrollOptions, PresetStore, ProjectModel, ProjectModelConfig, QuickFind, QuickFindConfig, RecordPositionContext, RecurrenceConfirmationPopup, RecurrenceConfirmationPopupConfig, RegionResize, RegionResizeConfig, ResourceAllocationInterval, ResourceHistogramRenderData, ResourceMarginConfig, ResourceMenu, ResourceMenuConfig, ResourceNonWorkingTime, ResourceNonWorkingTimeConfig, ResourceTimeRangeModel, ResourceTimeRangeModelConfig, ResourceTimeRangeStore, ResourceTimeRangeStoreConfig, ResourceTimeRanges, ResourceTimeRangesConfig, ResourceUtilization, ResourceUtilizationListeners, ResourceUtilizationModel, ResourceUtilizationRenderData, Row, RowCopyPaste, RowCopyPasteConfig, RowExpander, RowExpanderConfig, RowReorder, RowReorderConfig, ScaleColumnConfig, ScalePoint, ScheduleTooltip, ScheduleTooltipConfig, Scheduler, SchedulerEventModel, SchedulerPro, SchedulerProCellEdit, SchedulerProCellEditConfig, SchedulerProProjectModel, SchedulerResourceModel, SchedulerSummary, SchedulerSummaryConfig, SchedulerTimeAxisColumn, ScrollManager, ScrollManagerConfig, Scroller, ScrollerConfig, Search, SearchConfig, Sort, SortConfig, Split, SplitConfig, StateProvider, StickyCells, StickyCellsConfig, Store, StoreConfig, Stripe, StripeConfig, SubGrid, SubGridConfig, TabConfig, TimeAxis, TimeAxisConfig, TimeAxisHeaderMenu, TimeAxisHeaderMenuConfig, TimeRanges, TimeRangesConfig, TimeSpan, TimeSpanConfig, TimeSpanHighlight, TimeSpanHighlightConfig, TimelineBase, TimelineContext, TimelineHistogram, Tool, ToolConfig, ToolbarConfig, TooltipConfig, Tree, TreeConfig, VerticalTimeAxisColumn, ViewPreset, ViewPresetConfig, VisibleDate, Widget } from '@bryntum/gantt';

import { createWidget, shouldComponentUpdate, processWidgetContent } from './WrapperHelper.js';

export type BryntumResourceUtilizationProps = {
    // Configs
    /**
     * Element (or element id) to adopt as this Widget's encapsulating element. The widget's
     * content will be placed inside this element.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-adopt)
     */
    adopt? : HTMLElement|string
    /**
     * A function used for aggregating child records histogram data entries to their parent entry.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-aggregateDataEntry)
     * @param {object} aggregateDataEntry.aggregated Target parent data entry to aggregate the entry into.
     * @param {object} aggregateDataEntry.entry Current entry to aggregate into `aggregated`.
     * @param {number} aggregateDataEntry.arrayIndex Index of current array (index of the record among other records being aggregated).
     * @param {<Object>} aggregateDataEntry.entryIndex Index of `entry` in the current array.
     * @returns {object} Return value becomes the value of the `aggregated` parameter on the next invocation of this function.
     */
    aggregateDataEntry? : ((aggregateDataEntry: { aggregated: object, entry: object, arrayIndex: number, entryIndex: object[] }) => object)|string
    /**
     * When `true` the component will automatically calculate data for group records
     * based on the groups members data by calling [getGroupRecordHistogramData](https://bryntum.com/products/gantt/docs/api/Scheduler/view/mixin/TimelineHistogramGrouping#function-getGroupRecordHistogramData) method.
     */
    aggregateHistogramDataForGroups? : boolean
    /**
     * When this widget is a child of a [Container](https://bryntum.com/products/gantt/docs/api/Core/widget/Container), it will by default be participating in a
     * flexbox layout. This config allows you to set this widget's
     * [align-self](https://developer.mozilla.org/en-US/docs/Web/CSS/align-self) style.
     */
    alignSelf? : string
    /**
     * Set to `false` to prevent remove row animation and remove the delay related to that.
     */
    animateRemovingRows? : boolean
    /**
     * When the [Tree](https://bryntum.com/products/gantt/docs/api/Grid/feature/Tree) feature is in use and the Store is a tree store, this
     * config may be set to `true` to visually animate branch node expand and collapse operations.
     * *This is not supported in Scheduler and Gantt*
     */
    animateTreeNodeToggle? : boolean
    /**
     * Element (or the id of an element) to append this widget's element to. Can be configured, or set once at
     * runtime. To access the element of a rendered widget, see [element](https://bryntum.com/products/gantt/docs/api/Core/widget/Widget#property-element).
     */
    appendTo? : HTMLElement|string
    /**
     * A localizable string (May contain `'L{}'` tokens which resolve in the locale file) to inject
     * into an element which will be linked using the `aria-describedby` attribute.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-ariaDescription)
     */
    ariaDescription? : string
    /**
     * A localizable string (May contain `'L{}'` tokens which resolve in the locale file) to inject as
     * the `aria-label` attribute.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-ariaLabel)
     */
    ariaLabel? : string
    /**
     * The `aria-live` value for this widget.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-ariaLive)
     */
    ariaLive? : string
    /**
     * You can set this option to `false` to make the timeline panel start and end on the exact provided
     * [startDate](https://bryntum.com/products/gantt/docs/api/Scheduler/view/TimelineBase#config-startDate)/[endDate](#Scheduler/view/TimelineBase#config-endDate) w/o adjusting them.
     */
    autoAdjustTimeAxis? : boolean
    /**
     * Automatically set grids height to fit all rows (no scrolling in the grid). In general you should avoid
     * using `autoHeight: true`, since it will bypass Grids virtual rendering and render all rows at once, which
     * in a larger grid is really bad for performance.
     */
    autoHeight? : boolean
    /**
     * Controls how much space to leave between stacked event bars in px.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-barMargin)
     */
    barMargin? : number
    /**
     * Default time unit used for displaying resources effort in bars.
     * Yet the effective time unit used might change dynamically when zooming in the histogram
     * so its ticks unit gets smaller than the default unit.
     * Please use [barTipEffortUnit](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceHistogram#config-barTipEffortUnit) to customize default units for tooltips
     * (or [effortUnit](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceHistogram#config-effortUnit) to customize both texts and tooltips default units).
     */
    barTextEffortUnit? : DurationUnit
    /**
     * Default time unit used when displaying resources effort in tooltips.
     * Yet the effective time unit used might change dynamically when zooming in the histogram
     * so its ticks unit gets smaller than the default unit.
     * Please use [barTextEffortUnit](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceHistogram#config-barTextEffortUnit) to customize default units for bar texts
     * (or [effortUnit](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceHistogram#config-effortUnit) to customize both texts and tooltips default units).
     */
    barTipEffortUnit? : DurationUnit
    /**
     * A Function which returns the tooltip text to display when hovering a bar.
     * The following parameters are passed:
     * @param {object} data The backing data of the histogram rectangle
     * @param {object} data.rectConfig The rectangle configuration object
     * @param {object} data.datum The hovered bar data
     * @param {number} data.index The index of the hovered bar data
     * @param {ResourceUtilizationModel} data.record The record which effort the hovered bar displays.
     * @returns {string} Tooltip HTML content
     */
    barTooltipTemplate? : (data: { rectConfig: object, datum: object, index: number, record: ResourceUtilizationModel }) => string
    /**
     * A Config object representing the configuration of a [Toolbar](https://bryntum.com/products/gantt/docs/api/Core/widget/Toolbar),
     * or array of config objects representing the child items of a Toolbar. Another way to add a bbar is to use [strips](https://bryntum.com/products/gantt/docs/api/Core/widget/Panel#config-strips).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-bbar)
     */
    bbar? : (ContainerItemConfig|string)[]|ToolbarConfig|PagingToolbarConfig
    /**
     * Custom CSS classes to add to the panel's body element.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-bodyCls)
     */
    bodyCls? : string|object
    /**
     * An object where property names with a truthy value indicate which events should bubble up the ownership
     * hierarchy when triggered.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-bubbleEvents)
     */
    bubbleEvents? : object
    /**
     * This config defines the size of the start and end invisible parts of the timespan when [infiniteScroll](https://bryntum.com/products/gantt/docs/api/Scheduler/view/mixin/TimelineScroll#config-infiniteScroll) set to `true`.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-bufferCoef)
     */
    bufferCoef? : number
    /**
     * This config defines the scroll limit, which, when exceeded will cause a timespan shift.
     * The limit is calculated as the `panelWidth * [bufferCoef](https://bryntum.com/products/gantt/docs/api/Scheduler/view/mixin/TimelineScroll#config-bufferCoef) * bufferThreshold`. During scrolling, if the left or right side
     * has less than that of the rendered content - a shift is triggered.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-bufferThreshold)
     */
    bufferThreshold? : number
    /**
     * Set to `false` to not call onXXX method names (e.g. `onShow`, `onClick`), as an easy way to listen for events.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-callOnFunctions)
     */
    callOnFunctions? : boolean
    /**
     * Custom CSS classes to add to element.
     * May be specified as a space separated string, or as an object in which property names
     * with truthy values are used as the class names:
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-cls)
     */
    cls? : string|object
    /**
     * Controls whether the panel is collapsed (the body of the panel is hidden while only the header is
     * visible). Only valid if the panel is [collapsible](https://bryntum.com/products/gantt/docs/api/Core/widget/Panel#config-collapsible).
     */
    collapsed? : boolean
    /**
     * This config enables collapsibility for the panel. See [collapsed](https://bryntum.com/products/gantt/docs/api/Core/widget/Panel#config-collapsed).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-collapsible)
     */
    collapsible? : boolean|PanelCollapserConfig|PanelCollapserOverlayConfig
    /**
     * Set to `false` to not show column lines. End result might be overruled by/differ between themes.
     */
    columnLines? : boolean
    /**
     * Accepts column definitions for the grid during initialization. They will be used to create
     * [Column](https://bryntum.com/products/gantt/docs/api/Grid/column/Column) instances that are added to a [ColumnStore](#Grid/data/ColumnStore).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-columns)
     */
    columns? : ColumnStore|GridColumnConfig[]|ColumnStoreConfig
    config? : object
    /**
     * Custom CSS classes to add to the [contentElement](https://bryntum.com/products/gantt/docs/api/Core/widget/Widget#property-contentElement).
     * May be specified as a space separated string, or as an object in which property names
     * with truthy values are used as the class names:
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-contentElementCls)
     */
    contentElementCls? : string|object
    /**
     * Event which is used to show context menus.
     * Available options are: 'contextmenu', 'click', 'dblclick'.
     */
    contextMenuTriggerEvent? : 'contextmenu'|'click'|'dblclick'
    /**
     * Record field from which the histogram data will be collected.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-dataModelField)
     */
    dataModelField? : string
    /**
     * Object to apply to elements dataset (each key will be used as a data-attribute on the element)
     */
    dataset? : Record<string, string>
    /**
     * Region to which columns are added when they have none specified
     */
    defaultRegion? : string
    /**
     * Generic resource image, used when provided `imageUrl` or `image` fields or path calculated from resource
     * name are all invalid. If left blank, resource name initials will be shown when no image can be loaded.
     */
    defaultResourceImageName? : string
    /**
     * Set to `true` to destroy the store when the grid is destroyed.
     */
    destroyStore? : boolean
    /**
     * Configure as `true` to destroy the Project and stores when `this` is destroyed.
     */
    destroyStores? : boolean
    /**
     * Disable or enable the widget. It is similar to [readOnly](https://bryntum.com/products/gantt/docs/api/Core/widget/Widget#config-readOnly) except a disabled widget
     * cannot be focused, uses a different rendition (usually greyish) and does not allow selecting its value.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-disabled)
     */
    disabled? : boolean|'inert'
    /**
     * Set to `true` to not get a warning when calling [getState](https://bryntum.com/products/gantt/docs/api/Grid/view/GridBase#function-getState) when there is a column
     * configured without an `id`. But the recommended action is to always configure columns with an `id` when
     * using states.
     */
    disableGridColumnIdWarning? : boolean
    /**
     * Set to `true` to not get a warning when using another base class than GridRowModel for your grid data. If
     * you do, and would like to use the full feature set of the grid then include the fields from GridRowModel
     * in your model definition.
     */
    disableGridRowModelWarning? : boolean
    /**
     * Defines how dates will be formatted in tooltips etc. This config has priority over similar config on the
     * view preset. For allowed values see [format](https://bryntum.com/products/gantt/docs/api/Core/helper/DateHelper#function-format-static).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-displayDateFormat)
     */
    displayDateFormat? : string
    /**
     * Controls the placement of this widget when it is added to a [panel's ](https://bryntum.com/products/gantt/docs/api/Core/widget/Panel)
     * [strips collection](https://bryntum.com/products/gantt/docs/api/Core/widget/Panel#config-strips). Typical values for this config are `'top'`,
     * `'bottom'`, `'left'`, or `'right'`, which cause the widget to be placed on that side of the panel's
     * body. Such widgets are called "edge strips".
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-dock)
     */
    dock? : 'top'|'bottom'|'left'|'right'|'start'|'end'|'header'|'pre-header'|object
    /**
     * Effort value format string.
     * Must be a template supported by [NumberFormat](https://bryntum.com/products/gantt/docs/api/Core/helper/util/NumberFormat) class.
     */
    effortFormat? : string
    /**
     * Default time unit to display resources effort values.
     * The value is used as default when displaying effort in tooltips and bars text.
     * Yet the effective time unit used might change dynamically when zooming in the histogram
     * so its ticks unit gets smaller than the default unit.
     * Please use [barTipEffortUnit](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceHistogram#config-barTipEffortUnit) to customize default units for tooltips only
     * and [barTextEffortUnit](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceHistogram#config-barTextEffortUnit) to customize default units in bar texts.
     */
    effortUnit? : DurationUnit
    /**
     * Text or HTML, or a [DomConfig](https://bryntum.com/products/gantt/docs/api/Core/helper/DomHelper#typedef-DomConfig) block to display when there is no data to display in the grid
     */
    emptyText? : string|DomConfig
    /**
     * Set to `false` if you don't want event bar DOM updates to animate.
     */
    enableEventAnimations? : boolean
    /**
     * Configure this as `true` to allow elements within cells to be styled as `position: sticky`.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-enableSticky)
     */
    enableSticky? : boolean
    /**
     * Set to `true` to allow text selection in the grid cells. Note, this cannot be used simultaneously with the
     * `RowReorder` feature.
     */
    enableTextSelection? : boolean
    /**
     * Set to `true` to listen for CTRL-Z (CMD-Z on Mac OS) keyboard event and trigger undo (redo when SHIFT is
     * pressed). Only applicable when using a [StateTrackingManager](https://bryntum.com/products/gantt/docs/api/Core/data/stm/StateTrackingManager).
     */
    enableUndoRedoKeys? : boolean
    /**
     * The end date of the timeline (if not configure with [infiniteScroll](https://bryntum.com/products/gantt/docs/api/Scheduler/view/TimelineBase#config-infiniteScroll)).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-endDate)
     */
    endDate? : Date|string
    /**
     * The name of the end date parameter that will be passed to in every `eventStore` load request.
     */
    endParamName? : string
    extraData? : any
    /**
     * Set to `true` to stretch the last column in a grid with all fixed width columns
     * to fill extra available space if the grid's width is wider than the sum of all
     * configured column widths.
     */
    fillLastColumn? : boolean
    /**
     * Specify `true` to force rendered events/tasks to fill entire ticks. This only affects rendering, start
     * and end dates retain their value on the data level.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-fillTicks)
     */
    fillTicks? : boolean
    /**
     * Use fixed row height. Setting this to `true` will configure the underlying RowManager to use fixed row
     * height, which sacrifices the ability to use rows with variable height to gain a fraction better
     * performance.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-fixedRowHeight)
     */
    fixedRowHeight? : boolean
    /**
     * When this widget is a child of a [Container](https://bryntum.com/products/gantt/docs/api/Core/widget/Container), it will by default be participating in a
     * flexbox layout. This config allows you to set this widget's
     * [flex](https://developer.mozilla.org/en-US/docs/Web/CSS/flex) style.
     * This may be configured as a single number or a `&lt;flex-grow&gt; &lt;flex-shrink&gt; &lt;flex-basis&gt;` format string.
     * numeric-only values are interpreted as the `flex-grow` value.
     */
    flex? : number|string
    /**
     * Config object of a footer. May contain a `dock`, `html` and a `cls` property. A footer is not a widget,
     * but rather plain HTML that follows the last element of the panel's body and [strips](https://bryntum.com/products/gantt/docs/api/Core/widget/Panel#config-strips).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-footer)
     */
    footer? : {
        dock?: 'top'|'right'|'bottom'|'left'|'start'|'end'
        html?: string
        cls?: string
    }|string
    /**
     * Set to `true` to force the time columns to fit to the available space (horizontal or vertical depends on mode).
     * Note that setting [suppressFit](https://bryntum.com/products/gantt/docs/api/Scheduler/view/TimelineBase#config-suppressFit) to `true`, will disable `forceFit` functionality. Zooming
     * cannot be used when `forceFit` is set.
     */
    forceFit? : boolean
    /**
     * Refresh entire row when a record changes (`true`) or, if possible, only the cells affected (`false`).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-fullRowRefresh)
     */
    fullRowRefresh? : boolean
    /**
     * A Function which returns a CSS class name to add to a rectangle element.
     * The following parameters are passed:
     * @param {HistogramSeries} series The series being rendered
     * @param {DomConfig} domConfig The rectangle configuration object
     * @param {ResourceAllocationInterval|AssignmentAllocationInterval} datum The datum being rendered. Either [ResourceAllocationInterval](https://bryntum.com/products/gantt/docs/api/SchedulerPro/model/ResourceModel#typedef-ResourceAllocationInterval) object for resource records (root level records) or [AssignmentAllocationInterval](https://bryntum.com/products/gantt/docs/api/SchedulerPro/model/ResourceModel#typedef-AssignmentAllocationInterval) object for assignment records
     * @param {number} index The index of the datum being rendered
     * @param {ResourceUtilizationRenderData} renderData Current render data giving access to the record, row and cell being rendered.
     * @returns {string} CSS classes of the rectangle element
     */
    getBarClass? : (series: HistogramSeries, domConfig: DomConfig, datum: ResourceAllocationInterval|AssignmentAllocationInterval, index: number, renderData: ResourceUtilizationRenderData) => string
    /**
     * A Function which if provided returns DOM configuration object for a bar
     * (a `RECT` element representing a single "bar"-type value).
     * The function accepts default prepared DOM configuration
     * in an argument which then can be processed and returned.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-getBarDOMConfig)
     * @param {HistogramSeries} series The series being rendered
     * @param {DomConfig} domConfig The rectangle DOM configuration object
     * @param {ResourceAllocationInterval|AssignmentAllocationInterval} datum The datum being rendered. Either [ResourceAllocationInterval](https://bryntum.com/products/gantt/docs/api/SchedulerPro/model/ResourceModel#typedef-ResourceAllocationInterval) object for resource records (root level records) or [AssignmentAllocationInterval](https://bryntum.com/products/gantt/docs/api/SchedulerPro/model/ResourceModel#typedef-AssignmentAllocationInterval) object for assignment records
     * @param {number} index The index of the datum being rendered
     * @param {ResourceUtilizationRenderData} renderData Current render data giving access to the record, row and cell being rendered.
     * @returns {object} Resulting DOM configuration object. If no value returned the bar is not displayed.
     */
    getBarDOMConfig? : (series: HistogramSeries, domConfig: DomConfig, datum: ResourceAllocationInterval|AssignmentAllocationInterval, index: number, renderData: ResourceUtilizationRenderData) => object
    /**
     * A Function which returns the text to render inside a bar.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-getBarText)
     * @param {ResourceAllocationInterval|AssignmentAllocationInterval} datum The datum being rendered. Either [ResourceAllocationInterval](https://bryntum.com/products/gantt/docs/api/SchedulerPro/model/ResourceModel#typedef-ResourceAllocationInterval) object for resource records (root level records) or [AssignmentAllocationInterval](https://bryntum.com/products/gantt/docs/api/SchedulerPro/model/ResourceModel#typedef-AssignmentAllocationInterval) object for assignment records
     * @param {number} index The index of the datum being rendered
     * @returns {string} Text to render inside the bar
     */
    getBarText? : (datum: ResourceAllocationInterval|AssignmentAllocationInterval, index: number) => string
    /**
     * A Function which returns a DOM configuration object for text elements.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-getBarTextDOMConfig)
     * @param {DomConfig} domConfig The `TEXT` element DOM configuration object
     * @param {ResourceAllocationInterval|AssignmentAllocationInterval} datum The datum being rendered. Either [ResourceAllocationInterval](https://bryntum.com/products/gantt/docs/api/SchedulerPro/model/ResourceModel#typedef-ResourceAllocationInterval) object for resource records (root level records) or [AssignmentAllocationInterval](https://bryntum.com/products/gantt/docs/api/SchedulerPro/model/ResourceModel#typedef-AssignmentAllocationInterval) object for assignment records
     * @param {number} index The index of the datum being rendered
     * @param {HistogramSeries} series The series (provided if histogram widget [singleTextForAllBars](https://bryntum.com/products/gantt/docs/api/Core/widget/graph/Histogram#config-singleTextForAllBars) is `false`)
     * @param {ResourceUtilizationRenderData} renderData Current render data giving access to the record, row and cell being rendered.
     * @returns {object} The `TEXT` element DOM configuration object. If no value returned the text element is not displayed.
     */
    getBarTextDOMConfig? : (domConfig: DomConfig, datum: ResourceAllocationInterval|AssignmentAllocationInterval, index: number, series: HistogramSeries, renderData: ResourceUtilizationRenderData) => object
    /**
     * Function that extracts a record histogram data entry for aggregating.
     * By default it returns the entry as is. Override the function if you need a more complex way
     * to retrieve the value for aggregating.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-getDataEntryForAggregating)
     * @param {object} getDataEntryForAggregating.entry Current data entry.
     * @returns {object} Entry to aggregate
     */
    getDataEntryForAggregating? : ((getDataEntryForAggregating: { entry: object }) => object)|string
    /**
     * A Function which returns a CSS class name to add to a path element
     * built for an `outline` type series.
     * The following parameters are passed:
     * @param {HistogramSeries} series The series being rendered
     * @param {<Object>} data The series data
     * @param {ResourceHistogramRenderData} renderData Current render data giving access to the record, row and cell being rendered.
     * @returns {string} CSS class name of the path element
     */
    getOutlineClass? : (series: HistogramSeries, data: object[], renderData: ResourceHistogramRenderData) => string
    /**
     * A Function which if provided should return a DOM configuration object for
     * a `path` element built for an `outline` type series.
     * The function accepts a default prepared DOM configuration
     * in an argument which then can be processed and returned.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-getOutlineDOMConfig)
     * @param {HistogramSeries} series The series being rendered
     * @param {DomConfig} domConfig The `path` element DOM configuration object
     * @param {<Object>} data The series data
     * @param {ResourceHistogramRenderData} renderData Current render data giving access to the record, row and cell being rendered.
     * @returns {object} Resulting DOM configuration object. If no value returned the path is not displayed.
     */
    getOutlineDOMConfig? : (series: HistogramSeries, domConfig: DomConfig, data: object[], renderData: ResourceHistogramRenderData) => object
    /**
     * A function, or name of a function which builds histogram data for the provided record.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-getRecordData)
     * @param {Model} getRecordData.record Record to get histogram data for.
     * @param {object} aggregationContext Context object passed in case the data is being retrieved as a part of some parent record data collecting.
     * @returns {object} Histogram data.
     */
    getRecordData? : ((getRecordData: { record: Model }, aggregationContext?: object) => object)|string
    /**
     * A function called for each row to determine its height. It is passed a [record](https://bryntum.com/products/gantt/docs/api/Core/data/Model) and
     * expected to return the desired height of that records row. If the function returns a falsy value, Grids
     * configured [rowHeight](https://bryntum.com/products/gantt/docs/api/Grid/view/GridBase#config-rowHeight) is used.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-getRowHeight)
     * @param {Model} getRowHeight.record Record to determine row height for
     * @returns {number} Desired row height
     */
    getRowHeight? : (getRowHeight: { record: Model }) => number
    /**
     * When set to `true` (default) the component reacts on time axis changes
     * (zooming or changing the displayed time span), clears the histogram data cache of the records
     * and then refreshes the view.
     */
    hardRefreshOnTimeAxisReconfigure? : boolean
    /**
     * A config [object](https://bryntum.com/products/gantt/docs/api/Core/widget/Panel#typedef-PanelHeader) for the panel's header or a string in place of a `title`.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-header)
     */
    header? : string|boolean|PanelHeader
    /**
     * Widget's height, used to set element style.height. Either specify a valid height string or a number, which
     * will get 'px' appended. We recommend using CSS as the primary way to control height, but in some cases
     * this config is convenient.
     */
    height? : string|number
    /**
     * Configure with true to make widget initially hidden.
     */
    hidden? : boolean
    /**
     * Set to `true` to hide the footer elements
     */
    hideFooters? : boolean
    /**
     * Set to `true` to hide the column header elements
     */
    hideHeaders? : boolean
    /**
     * Set to `true` to hide the Grid's horizontal scrollbar(s)
     */
    hideHorizontalScrollbar? : boolean
    /**
     * By default the row hover effect is not visible in the Scheduler part of the grid.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-hideRowHover)
     */
    hideRowHover? : boolean
    /**
     * An instance or a configuration object of the underlying [Histogram](https://bryntum.com/products/gantt/docs/api/Core/widget/graph/Histogram)
     * component that is used under the hood to render actual charts.
     * In case a configuration object is provided the built class is defined with
     * [histogramWidgetClass](https://bryntum.com/products/gantt/docs/api/Scheduler/view/TimelineHistogramBase#config-histogramWidgetClass) config.
     */
    histogramWidget? : Histogram|HistogramConfig
    /**
     * The class used for building the [histogram widget](https://bryntum.com/products/gantt/docs/api/Scheduler/view/TimelineHistogramBase#property-histogramWidget)
     */
    histogramWidgetClass? : Histogram
    /**
     * A CSS class to add to hovered row elements
     */
    hoverCls? : string
    /**
     * Widget id, if not specified one will be generated. Also used for lookups through Widget.getById
     */
    id? : string
    /**
     * Set to `true` to ignore reacting to DOM events (mouseover/mouseout etc) while scrolling. Useful if you
     * want to maximize scroll performance.
     */
    ignoreDomEventsWhileScrolling? : boolean
    /**
     * Determines if the widgets read-only state should be controlled by its parent.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-ignoreParentReadOnly)
     */
    ignoreParentReadOnly? : boolean
    /**
     * Set to `true` to include inactive tasks allocation and `false` to not take such tasks into account.
     */
    includeInactiveEvents? : boolean
    /**
     * Configure as `true` to automatically adjust the panel timespan during scrolling in the time dimension,
     * when the scroller comes close to the start/end edges.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-infiniteScroll)
     */
    infiniteScroll? : boolean
    /**
     * A function that initializes a target group record entry.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-initAggregatedDataEntry)
     * @returns {object} Target aggregated entry
     */
    initAggregatedDataEntry? : (() => object)|string
    /**
     * Element (or element id) to insert this widget before. If provided, [appendTo](https://bryntum.com/products/gantt/docs/api/Core/widget/Widget#config-appendTo) config is ignored.
     */
    insertBefore? : HTMLElement|string
    /**
     * Element (or element id) to append this widget element to, as a first child. If provided, [appendTo](https://bryntum.com/products/gantt/docs/api/Core/widget/Widget#config-appendTo) config is ignored.
     */
    insertFirst? : HTMLElement|string
    /**
     * See [Keyboard shortcuts](https://bryntum.com/products/gantt/docs/api/Grid/view/Grid#keyboard-shortcuts) for details
     */
    keyMap? : Record<string, KeyMapConfig>
    /**
     * The listener set for this object.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-listeners)
     */
    listeners? : ResourceUtilizationListeners
    /**
     * A [Mask](https://bryntum.com/products/gantt/docs/api/Core/widget/Mask) config object, or a message to be shown when a store is performing a remote
     * operation, or Crud Manager is loading data from the sever. Set to `null` to disable default load mask.
     */
    loadMask? : string|MaskConfig|null
    /**
     * A [Mask](https://bryntum.com/products/gantt/docs/api/Core/widget/Mask) config object to adjust the [maskDefaults](#Core/widget/Widget#config-maskDefaults)
     * when data is loading. The message and optional configuration from the
     * [loadMask](https://bryntum.com/products/gantt/docs/api/Core/mixin/LoadMaskable#config-loadMask) config take priority over these options, just as they do
     * for `maskDefaults`, respectively.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-loadMaskDefaults)
     */
    loadMaskDefaults? : MaskConfig
    /**
     * A [Mask](https://bryntum.com/products/gantt/docs/api/Core/widget/Mask) config object to adjust the [maskDefaults](#Core/widget/Widget#config-maskDefaults)
     * when an error occurs loading data.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-loadMaskError)
     */
    loadMaskError? : MaskConfig|Mask|boolean
    /**
     * Set to `false` to disable localization of this object.
     */
    localizable? : boolean
    /**
     * Time in ms until a longpress is triggered
     */
    longPressTime? : number
    /**
     * When `true`, events are sized and positioned based on rowHeight, resourceMargin and barMargin settings.
     * Set this to `false` if you want to control height and vertical position using CSS instead.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-managedEventSizing)
     */
    managedEventSizing? : boolean
    /**
     * Widget's margin. This may be configured as a single number or a `TRBL` format string.
     * numeric-only values are interpreted as pixels.
     */
    margin? : number|string
    /**
     * Grids change the `maskDefaults` to cover only their `body` element.
     */
    maskDefaults? : MaskConfig
    /**
     * Set to `true` to apply the default mask to the widget. Alternatively, this can be the mask message or a
     * [Mask](https://bryntum.com/products/gantt/docs/api/Core/widget/Mask) config object.
     */
    masked? : boolean|string|MaskConfig
    /**
     * The element's maxHeight. Can be either a String or a Number (which will have 'px' appended). Note that
     * like [height](https://bryntum.com/products/gantt/docs/api/Core/widget/Widget#config-height), *reading* the value will return the numeric value in pixels.
     */
    maxHeight? : string|number
    /**
     * The maximum time axis unit to display non-working ranges for ('hour' or 'day' etc).
     * When zooming to a view with a larger unit, no non-working time elements will be rendered.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-maxTimeAxisUnit)
     */
    maxTimeAxisUnit? : DurationUnit
    /**
     * The elements maxWidth. Can be either a String or a Number (which will have 'px' appended). Note that
     * like [width](https://bryntum.com/products/gantt/docs/api/Core/widget/Widget#config-width), *reading* the value will return the numeric value in pixels.
     */
    maxWidth? : string|number
    /**
     * The maximum zoom level to which [zoomIn](https://bryntum.com/products/gantt/docs/api/Scheduler/view/mixin/TimelineZoomable#function-zoomIn) will work. Defaults to the number of
     * [ViewPresets](https://bryntum.com/products/gantt/docs/api/Scheduler/preset/ViewPreset) available, see [presets](#Scheduler/view/mixin/TimelineViewPresets#property-presets)
     * for information. Unless you have modified the collection of available presets, the max zoom level is
     * milliseconds.
     */
    maxZoomLevel? : number
    /**
     * Grid's `min-height`. Defaults to `10em` to be sure that the Grid always has a height wherever it is
     * inserted.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-minHeight)
     */
    minHeight? : string|number
    /**
     * The elements minWidth. Can be either a String or a Number (which will have 'px' appended). Note that
     * like [width](https://bryntum.com/products/gantt/docs/api/Core/widget/Widget#config-width), *reading* the value will return the numeric value in pixels.
     */
    minWidth? : string|number
    /**
     * The minimum zoom level to which [zoomOut](https://bryntum.com/products/gantt/docs/api/Scheduler/view/mixin/TimelineZoomable#function-zoomOut) will work. Defaults to 0 (year ticks)
     */
    minZoomLevel? : number
    /**
     * Grid monitors window resize by default.
     */
    monitorResize? : boolean
    /**
     * The owning Widget of this Widget. If this Widget is directly contained (that is, it is one of the
     * [items](https://bryntum.com/products/gantt/docs/api/Core/widget/Container#property-items) of a Container), this config will be ignored. In this case
     * the owner is <strong>always</strong> the encapsulating Container.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-owner)
     */
    owner? : Widget
    /**
     * Partners this Timeline panel with another Timeline in order to sync their region sizes (sub-grids like locked, normal will get the same width),
     * start and end dates, view preset, zoom level and scrolling position. All these values will be synced with the timeline defined as the `partner`.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-partner)
     */
    partner? : TimelineBase
    /**
     * Set to `true` to include `startDate` and `endDate` params indicating the currently viewed date range in
     * `EventStore` load requests (when loading using `AjaxStore` or `CrudManager` functionality).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-passStartEndParameters)
     */
    passStartEndParameters? : boolean
    /**
     * Specify plugins (an array of classes) in config
     */
    plugins? : Function[]
    /**
     * True to preserve focused cell after loading new data
     */
    preserveFocusOnDatasetChange? : boolean
    /**
     * Preserve the grid's vertical scroll position when changesets are applied, as in the case of remote
     * changes, or when stores are configured with [syncDataOnLoad](https://bryntum.com/products/gantt/docs/api/Core/data/Store#config-syncDataOnLoad).
     */
    preserveScroll? : PreserveScrollOptions|boolean
    /**
     * Specify `true` to preserve vertical scroll position after store actions that trigger a `refresh` event,
     * such as loading new data and filtering.
     */
    preserveScrollOnDatasetChange? : boolean
    /**
     * An array of [ViewPreset](https://bryntum.com/products/gantt/docs/api/Scheduler/preset/ViewPreset) config objects
     * which describes the available timeline layouts for this scheduler.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-presets)
     */
    presets? : ViewPresetConfig[]
    /**
     * Prevent tooltip from being displayed on touch devices. Useful for example for buttons that display a
     * menu on click etc, since the tooltip would be displayed at the same time.
     */
    preventTooltipOnTouch? : boolean
    /**
     * A ProjectModel instance (or a config object) to display resource allocation of.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-project)
     */
    project? : ProjectModelConfig|SchedulerProProjectModel|RefObject<any>
    /**
     * Set to `true` to make the grid read-only, by disabling any UIs for modifying data.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-readOnly)
     */
    readOnly? : boolean
    /**
     * The confirmation dialog shown when a recurring event is edited.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-recurrenceConfirmationPopup)
     */
    recurrenceConfirmationPopup? : RecurrenceConfirmationPopupConfig|RecurrenceConfirmationPopup
    relayStoreEvents? : boolean
    /**
     * Set to `false` to only measure cell contents when double-clicking the edge between column headers.
     */
    resizeToFitIncludesHeader? : boolean
    /**
     * Resource image extension, used when creating image path from resource name.
     */
    resourceImageExtension? : string
    /**
     * Path to load resource images from. Used by the resource header in vertical mode and the
     * [ResourceInfoColumn](https://bryntum.com/products/gantt/docs/api/Scheduler/column/ResourceInfoColumn) in horizontal mode. Set this to display miniature
     * images for each resource using their `image` or `imageUrl` fields.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-resourceImagePath)
     */
    resourceImagePath? : string
    /**
     * Control how much space to leave between the first event/last event and the resources edge (top/bottom
     * margin within the resource row in horizontal mode, left/right margin within the resource column in
     * vertical mode), in px. Defaults to the value of [barMargin](https://bryntum.com/products/gantt/docs/api/Scheduler/view/Scheduler#config-barMargin).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-resourceMargin)
     */
    resourceMargin? : number|ResourceMarginConfig
    /**
     * An empty function by default, but provided so that you can override it. This function is called each time
     * a resource time range is rendered into the schedule. It's called with `resourceTimeRangeRecord`, `resourceRecord`,
     * and `renderData` params.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-resourceTimeRangeRenderer)
     * @param {object} detail An object containing the information needed to render a ResourceTimeRangeModel.
     * @param {ResourceTimeRangeModel} detail.resourceTimeRangeRecord The resource time range record.
     * @param {ResourceModel} detail.resourceRecord The resource record.
     * @param {object} detail.renderData An object containing details about the event rendering.
     * @returns {string|DomConfig|DomConfig>} A simple string, or a DomConfig object describing the markup
     */
    resourceTimeRangeRenderer? : (detail: { resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, renderData: object }) => string|DomConfig|DomConfig[]
    /**
     * Inline resource time ranges, will be loaded into an internally created store if
     * [ResourceTimeRanges](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/ResourceTimeRanges) is enabled.
     */
    resourceTimeRanges? : ResourceTimeRangeModel[]|ResourceTimeRangeModelConfig[]
    /**
     * Resource time ranges store instance or config object for [ResourceTimeRanges](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/ResourceTimeRanges) feature.
     */
    resourceTimeRangeStore? : ResourceTimeRangeStore|ResourceTimeRangeStoreConfig
    /**
     * "Break points" for which responsive config to use for columns and css.
     */
    responsiveLevels? : Record<string, number|string>
    /**
     * Configure as `true` to have the component display a translucent ripple when its
     * [focusElement](https://bryntum.com/products/gantt/docs/api/Core/widget/Widget#property-focusElement), or [element](#Core/widget/Widget#property-element) is tapped <em>if the
     * current theme supports ripples</em>. Out of the box, only the Material theme supports ripples.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-ripple)
     */
    ripple? : boolean|{
        delegate?: string
        color?: string
        radius?: number
        clip?: string
    }
    /**
     * If you are rendering this widget to a shadow root inside a web component, set this config to the shadowRoot. If not inside a web component, set it to `document.body`
     */
    rootElement? : ShadowRoot|HTMLElement
    /**
     * Row height in pixels. This allows the default height for rows to be controlled. Note that it may be
     * overriden by specifying a [rowHeight](https://bryntum.com/products/gantt/docs/api/Grid/data/GridRowModel#field-rowHeight) on a per record basis, or from
     * a column [renderer](https://bryntum.com/products/gantt/docs/api/Grid/column/Column#config-renderer).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-rowHeight)
     */
    rowHeight? : number
    /**
     * Set to `false` to not show row lines. End result might be overruled by/differ between themes.
     */
    rowLines? : boolean
    /**
     * An object with configuration for the [ScaleColumn](https://bryntum.com/products/gantt/docs/api/Scheduler/column/ScaleColumn).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-scaleColumn)
     */
    scaleColumn? : ScaleColumnConfig
    /**
     * Array of objects representing the [scale column](https://bryntum.com/products/gantt/docs/api/Scheduler/view/mixin/TimelineHistogramScaleColumn#config-scaleColumn) scale points.
     * The config basically is a mapping to the column [scalePoints](https://bryntum.com/products/gantt/docs/api/Scheduler/column/ScaleColumn#config-scalePoints)
     * config.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-scalePoints)
     */
    scalePoints? : ScalePoint[]
    /**
     * Configures whether the grid is scrollable in the `Y` axis. This is used to configure a [Scroller](https://bryntum.com/products/gantt/docs/api/Core/helper/util/Scroller).
     * See the [scrollerClass](https://bryntum.com/products/gantt/docs/api/Grid/view/GridBase#config-scrollerClass) config option.
     */
    scrollable? : boolean|ScrollerConfig|Scroller
    /**
     * The class to instantiate to use as the [scrollable](https://bryntum.com/products/gantt/docs/api/Grid/view/GridBase#config-scrollable). Defaults to [Scroller](#Core/helper/util/Scroller).
     */
    scrollerClass? : typeof Scroller
    /**
     * Configuration values for the [ScrollManager](https://bryntum.com/products/gantt/docs/api/Core/util/ScrollManager) class on initialization. Returns the
     * [ScrollManager](https://bryntum.com/products/gantt/docs/api/Core/util/ScrollManager) at runtime.
     */
    scrollManager? : ScrollManagerConfig|ScrollManager
    /**
     * Selection configuration settings, change these properties to control how selection works and what can be
     * selected.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-selectionMode)
     */
    selectionMode? : GridSelectionMode
    /**
     * Object enumerating data series for the histogram.
     * The object keys are treated as the series identifiers and values are objects that
     * must contain two properties:
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-series)
     */
    series? : Record<string, HistogramSeries>
    /**
     * Set to `true` if you want to display resources effort values in bars
     * (for example: `24h`, `7d`, `60min` etc.).
     * The text contents can be changed by providing [getBarText](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-getBarText) function.
     */
    showBarText? : boolean
    /**
     * Set to `true` if you want to display a tooltip when hovering an allocation bar. You can also pass a
     * [Tooltip#configs](https://bryntum.com/products/gantt/docs/api/Core/widget/Tooltip#configs) config object.
     * Please use [barTooltipTemplate](https://bryntum.com/products/gantt/docs/api/Scheduler/view/TimelineHistogramBase#config-barTooltipTemplate) function to customize the tooltip contents.
     */
    showBarTip? : boolean|TooltipConfig
    /**
     * Configure as `true` to have the grid show a red "changed" tag in cells who's
     * field value has changed and not yet been committed.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-showDirty)
     */
    showDirty? : boolean|{
        duringEdit?: boolean
    }
    /**
     * Specifies whether effort values should display units or not.
     */
    showEffortUnit? : boolean
    /**
     * Set to `true` to snap to the current time resolution increment while interacting with scheduled events.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-snap)
     */
    snap? : boolean
    /**
     * Affects drag drop and resizing of events when [snap](https://bryntum.com/products/gantt/docs/api/Scheduler/view/mixin/TimelineDateMapper#config-snap)
     * is enabled.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-snapRelativeToEventStartDate)
     */
    snapRelativeToEventStartDate? : boolean
    /**
     * The start date of the timeline (if not configure with [infiniteScroll](https://bryntum.com/products/gantt/docs/api/Scheduler/view/TimelineBase#config-infiniteScroll)).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-startDate)
     */
    startDate? : Date|string
    /**
     * The name of the start date parameter that will be passed to in every `eventStore` load request.
     */
    startParamName? : string
    /**
     * This value can be one of the following:
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-stateful)
     */
    stateful? : boolean|object|string[]
    /**
     * The events that, when fired by this component, should trigger it to save its state by calling
     * [saveState](https://bryntum.com/products/gantt/docs/api/Core/mixin/State#function-saveState).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-statefulEvents)
     */
    statefulEvents? : object|string[]
    /**
     * The key to use when saving this object's state in the [stateProvider](https://bryntum.com/products/gantt/docs/api/Core/mixin/State#config-stateProvider). If this config is
     * not assigned, and [stateful](https://bryntum.com/products/gantt/docs/api/Core/mixin/State#config-stateful) is not set to `false`, the [id](#Core/widget/Widget#config-id)
     * (if explicitly specified) will be used as the `stateId`.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-stateId)
     */
    stateId? : string
    /**
     * The `StateProvider` to use to save and restore this object's [state](https://bryntum.com/products/gantt/docs/api/Core/mixin/State#property-state). By default, `state`
     * will be saved using the [default state provider](https://bryntum.com/products/gantt/docs/api/Core/state/StateProvider#property-instance-static).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-stateProvider)
     */
    stateProvider? : StateProvider
    /**
     * When set, the text in the major time axis header sticks in the scrolling viewport as long as possible.
     */
    stickyHeaders? : boolean
    /**
     * Store that holds records to display in the grid, or a store config object. If the configuration contains
     * a `readUrl`, an `AjaxStore` will be created.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-store)
     */
    store? : Store|StoreConfig|AjaxStore|AjaxStoreConfig
    /**
     * An object containing widgets keyed by name. By default (when no `type` is given), strips are
     * [toolbars](https://bryntum.com/products/gantt/docs/api/Core/widget/Toolbar). If you want to pass an array, you can use
     * the toolbar's [items](https://bryntum.com/products/gantt/docs/api/Core/widget/Container#config-items).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-strips)
     */
    strips? : Record<string, ContainerItemConfig>
    /**
     * An object containing sub grid configuration objects keyed by a `region` property.
     * By default, grid has a 'locked' region (if configured with locked columns) and a 'normal' region.
     * The 'normal' region defaults to use `flex: 1`.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-subGridConfigs)
     */
    subGridConfigs? : Record<string, SubGridConfig>
    /**
     * Set to `true` to prevent auto calculating of a minimal [tickSize](https://bryntum.com/products/gantt/docs/api/Scheduler/view/mixin/TimelineEventRendering#property-tickSize)
     * to always fit the content to the screen size. Setting this property on `true` will disable [forceFit](https://bryntum.com/products/gantt/docs/api/Scheduler/view/TimelineBase#config-forceFit) behaviour.
     */
    suppressFit? : boolean
    /**
     * A [Mask](https://bryntum.com/products/gantt/docs/api/Core/widget/Mask) config object, or a message to be shown when Crud Manager
     * is persisting changes on the server. Set to `null` to disable default sync mask.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-syncMask)
     */
    syncMask? : string|MaskConfig|null
    /**
     * A configuration for the [tab](https://bryntum.com/products/gantt/docs/api/Core/widget/Tab) created for this widget when it is placed in a
     * [TabPanel](https://bryntum.com/products/gantt/docs/api/Core/widget/TabPanel). For example, this config can be used to control the icon of the `tab` for
     * this widget:
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-tab)
     */
    tab? : boolean|TabConfig
    /**
     * A Config object representing the configuration of a [Toolbar](https://bryntum.com/products/gantt/docs/api/Core/widget/Toolbar),
     * or array of config objects representing the child items of a Toolbar.
     * This creates a toolbar docked to the top of the panel immediately below the header.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-tbar)
     */
    tbar? : (ContainerItemConfig|string)[]|ToolbarConfig|PagingToolbarConfig
    /**
     * The width/height (depending on vertical / horizontal mode) of all the time columns.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-tickSize)
     */
    tickSize? : number
    /**
     * A [TimeAxis](https://bryntum.com/products/gantt/docs/api/Scheduler/data/TimeAxis) config object or instance, used to create a backing data store of
     * 'ticks' providing the input date data for the time axis of timeline panel. Created automatically if none
     * supplied.
     */
    timeAxis? : TimeAxisConfig|TimeAxis
    /**
     * Inline time ranges, will be loaded into an internally created store if [TimeRanges](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/TimeRanges)
     * is enabled.
     */
    timeRanges? : TimeSpan[]|TimeSpanConfig[]
    /**
     * The time ranges store instance for [TimeRanges](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/TimeRanges) feature.
     */
    timeRangeStore? : Store|StoreConfig
    /**
     * A title to display in the header or owning TabPanel. Causes creation and docking of a header
     * to the top if no header is configured.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-title)
     */
    title? : string
    /**
     * The [tools](https://bryntum.com/products/gantt/docs/api/Core/widget/Tool) to add either before or after the `title` in the Panel header. Each
     * property name is the reference by which an instantiated tool may be retrieved from the live
     * `[tools](https://bryntum.com/products/gantt/docs/api/Core/widget/mixin/Toolable#property-tools)` property.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-tools)
     */
    tools? : Record<string, ToolConfig>
    /**
     * Animation transition duration in milliseconds.
     */
    transitionDuration? : number
    type? : 'resourceutilization'
    /**
     * Custom CSS class name suffixes to apply to the elements rendered by this widget. This may be specified
     * as a space separated string, an array of strings, or as an object in which property names with truthy
     * values are used as the class names.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-ui)
     */
    ui? : 'plain'|'toolbar'|string|object
    /**
     * A string key used to lookup a predefined [ViewPreset](https://bryntum.com/products/gantt/docs/api/Scheduler/preset/ViewPreset) (e.g. 'weekAndDay', 'hourAndDay'),
     * managed by [PresetManager](https://bryntum.com/products/gantt/docs/api/Scheduler/preset/PresetManager). See [PresetManager](#Scheduler/preset/PresetManager) for more information.
     * Or a config object for a viewPreset.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-viewPreset)
     */
    viewPreset? : string|ViewPresetConfig
    /**
     * A date to bring into view initially on the scrollable timeline.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-visibleDate)
     */
    visibleDate? : Date|VisibleDate
    /**
     * Integer number indicating the size of timespan during zooming. When zooming, the timespan is adjusted to make
     * the scrolling area `visibleZoomFactor` times wider than the timeline area itself. Used in
     * [zoomToSpan](https://bryntum.com/products/gantt/docs/api/Scheduler/view/mixin/TimelineZoomable#function-zoomToSpan) and [zoomToLevel](#Scheduler/view/mixin/TimelineZoomable#function-zoomToLevel) functions.
     */
    visibleZoomFactor? : number
    /**
     * A valid JS day index between 0-6 (0: Sunday, 1: Monday etc.) to be considered the start day of the week.
     * When omitted, the week start day is retrieved from the active locale class.
     */
    weekStartDay? : number
    /**
     * A widgets weight determines its position among siblings when added to a [Container](https://bryntum.com/products/gantt/docs/api/Core/widget/Container).
     * Higher weights go further down.
     */
    weight? : number
    /**
     * Widget's width, used to set element style.width. Either specify a valid width string or a number, which
     * will get 'px' appended. We recommend using CSS as the primary way to control width, but in some cases
     * this config is convenient.
     */
    width? : string|number
    /**
     * An object with format `{ fromDay, toDay, fromHour, toHour }` that describes the working days and hours.
     * This object will be used to populate TimeAxis [include](https://bryntum.com/products/gantt/docs/api/Scheduler/data/TimeAxis#config-include) property.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#config-workingTime)
     */
    workingTime? : {
        fromDay?: number
        toDay?: number
        fromHour?: number
        toHour?: number
    }
    /**
     * Whether the originally rendered timespan should be preserved while zooming. By default, it is set to `false`,
     * meaning the timeline panel will adjust the currently rendered timespan to limit the amount of HTML content to
     * render. When setting this option to `true`, be careful not to allow to zoom a big timespan in seconds
     * resolution for example. That will cause <strong>a lot</strong> of HTML content to be rendered and affect performance. You
     * can use [minZoomLevel](https://bryntum.com/products/gantt/docs/api/Scheduler/view/mixin/TimelineZoomable#config-minZoomLevel) and [maxZoomLevel](#Scheduler/view/mixin/TimelineZoomable#config-maxZoomLevel) config options for that.
     */
    zoomKeepsOriginalTimespan? : boolean
    /**
     * If true, you can zoom in and out on the time axis using CTRL-key + mouse wheel.
     */
    zoomOnMouseWheel? : boolean
    /**
     * True to zoom to time span when double-clicking a time axis cell.
     */
    zoomOnTimeAxisDoubleClick? : boolean

    // Events
    /**
     * Fires before an assignment is removed. Can be triggered by user pressing [DELETE] or [BACKSPACE] or
     * by the event editor. Can for example be used to display a custom dialog to confirm deletion, in which
     * case records should be "manually" removed after confirmation:
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-beforeAssignmentDelete)
     * @param {object} event Event object
     * @param {Scheduler} event.source The Scheduler instance
     * @param {SchedulerEventModel[]} event.eventRecords The records about to be deleted
     * @param {object} event.context Additional removal context:
     * @param {Function} event.context.finalize Function to call to finalize the removal. Used to asynchronously decide to remove the records or not. Provide `false` to the function to prevent the removal.
     * @param {boolean} event.context.finalize.removeRecords Provide `false` to the function to prevent the removal.
     */
    onBeforeAssignmentDelete? : ((event: { source: Scheduler, eventRecords: SchedulerEventModel[], context: { finalize: { removeRecords?: boolean } } }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Grid before the cell editing is canceled, return `false` to prevent cancellation.
     * @param {object} event Event object
     * @param {Grid} event.source Owner grid
     * @param {GridLocation} event.editorContext Editing context
     */
    onBeforeCancelCellEdit? : ((event: { source: Grid, editorContext: GridLocation }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Grid before deleting a range of selected cell values by pressing `Backspace` or `Del`
     * buttons while [autoEdit](https://bryntum.com/products/gantt/docs/api/Grid/feature/CellEdit#config-autoEdit) is set to `true`. Return `false` to prevent editing.
     * @param {object} event Event object
     * @param {Grid} event.source Owner grid
     * @param {(GridLocation|Model)[]} event.gridSelection An array of cell selectors or records that will have their values deleted (the records themself will not get deleted, only visible column values).
     */
    onBeforeCellDelete? : ((event: { source: Grid, gridSelection: (GridLocation|Model)[] }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Grid before editing starts, return `false` to prevent editing
     * @param {object} event Event object
     * @param {Grid} event.source Owner grid
     * @param {GridLocation} event.editorContext Editing context
     * @param {Column} event.editorContext.column Target column
     * @param {Model} event.editorContext.record Target record
     * @param {HTMLElement} event.editorContext.cell Target cell
     * @param {Field} event.editorContext.editor The input field that the column is configured with (see [field](https://bryntum.com/products/gantt/docs/api/Grid/column/Column#config-field)). This property mey be replaced to be a different [field](https://bryntum.com/products/gantt/docs/api/Core/widget/Field) in the handler, to take effect just for the impending edit.
     * @param {Function} event.editorContext.finalize An async function may be injected into this property which performs asynchronous finalization tasks such as complex validation of confirmation. The value `true` or `false` must be returned.
     * @param {object} event.editorContext.finalize.context An object describing the editing context upon requested completion of the edit.
     * @param {any} event.editorContext.value Cell value
     */
    onBeforeCellEditStart? : ((event: { source: Grid, editorContext: { column: Column, record: Model, cell: HTMLElement, editor: Field, finalize: { context?: object }, value: any } }) => Promise<boolean>|boolean|void)|string
    /**
     * This event is fired prior to starting a column drag gesture. The drag is canceled if a listener returns `false`.
     * @param {object} event Event object
     * @param {Grid} event.source The grid instance.
     * @param {Column} event.column The dragged column.
     * @param {Event} event.event The browser event.
     */
    onBeforeColumnDragStart? : ((event: { source: Grid, column: Column, event: Event }) => Promise<boolean>|boolean|void)|string
    /**
     * This event is fired when a column is dropped, and you can return false from a listener to abort the operation.
     * @param {object} event Event object
     * @param {Grid} event.source The grid instance.
     * @param {Column} event.column The dragged column.
     * @param {Column} event.insertBefore The column before which the dragged column will be inserted.
     * @param {Column} event.newParent The new parent column.
     * @param {Event} event.event The browser event.
     * @param {string} event.region The region where the column was dropped.
     */
    onBeforeColumnDropFinalize? : ((event: { source: Grid, column: Column, insertBefore: Column, newParent: Column, event: Event, region: string }) => Promise<boolean>|boolean|void)|string
    /**
     * This event is fired prior to starting a column resize gesture. The resize is canceled if a listener returns
     * `false`.
     * @param {object} event Event object
     * @param {Grid} event.source The grid instance
     * @param {Column} event.column The column
     * @param {Event} event.domEvent The browser event
     */
    onBeforeColumnResize? : ((event: { source: Grid, column: Column, domEvent: Event }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Grid before a copy action is performed, return `false` to prevent the action
     * @param {object} event Event object
     * @param {Grid} event.source Owner grid
     * @param {GridLocation[]} event.cells The cells about to be copied or cut
     * @param {string} event.data The string data about to be copied or cut
     * @param {boolean} event.isCut `true` if this is a cut action
     * @param {string} event.entityName 'cell' to distinguish this event from other beforeCopy events
     */
    onBeforeCopy? : ((event: { source: Grid, cells: GridLocation[], data: string, isCut: boolean, entityName: string }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Grid before CSV export starts. Return `false` to cancel the export.
     * @param {object} event Event object
     * @param {ExportConfig} event.config Export config
     * @param {Column[]} event.columns An array of columns to export
     * @param {Model[]} event.rows An array of records to export
     * @param {string} event.lineDelimiter The CSV delimiter to separate lines
     * @param {string} event.columnDelimiter The CSV delimiter to separate values on one line
     */
    onBeforeCSVExport? : ((event: { config: ExportConfig, columns: Column[], rows: Model[], lineDelimiter: string, columnDelimiter: string }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires before an object is destroyed.
     * @param {object} event Event object
     * @param {Base} event.source The Object that is being destroyed.
     */
    onBeforeDestroy? : ((event: { source: Base }) => void)|string
    /**
     * Fires before an event is removed. Can be triggered by user pressing [DELETE] or [BACKSPACE] or by the
     * event editor. Return `false` to immediately veto the removal (or a `Promise` yielding `true` or `false`
     * for async vetoing).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-beforeEventDelete)
     * @param {object} event Event object
     * @param {Scheduler|any} event.source The Scheduler instance
     * @param {SchedulerEventModel[]} event.eventRecords The records about to be deleted
     * @param {object} event.context Additional removal context:
     * @param {Function} event.context.finalize Function to call to finalize the removal. Used to asynchronously decide to remove the records or not. Provide `false` to the function to prevent the removal.
     * @param {boolean} event.context.finalize.removeRecords Provide `false` to the function to prevent the removal.
     */
    onBeforeEventDelete? : ((event: { source: Scheduler|any, eventRecords: SchedulerEventModel[], context: { finalize: { removeRecords?: boolean } } }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Grid before Excel export starts. Return `false` to cancel the export.
     * @param {object} event Event object
     * @param {ExportConfig} event.config Export config
     * @param {Column[]} event.columns An array of columns to export
     * @param {Model[]} event.rows An array of records to export
     */
    onBeforeExcelExport? : ((event: { config: ExportConfig, columns: Column[], rows: Model[] }) => Promise<boolean>|boolean|void)|string
    /**
     * Fired before dragging of the FillHandle starts, return `false` to prevent the drag operation.
     * @param {object} event Event object
     * @param {GridLocation} event.cell Information about the column / record
     * @param {MouseEvent} event.domEvent The raw DOM event
     */
    onBeforeFillHandleDragStart? : ((event: { cell: GridLocation, domEvent: MouseEvent }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Grid before the cell editing is finished, return false to signal that the value is invalid and editing should not be finalized.
     * @param {object} event Event object
     * @param {Grid} event.grid Target grid
     * @param {GridLocation} event.editorContext Editing context
     * @param {Editor} event.editorContext.editor The Editor being used. Will contain an `inputField` property which is the field being used to perform the editing.
     * @param {Column} event.editorContext.column Target column
     * @param {Model} event.editorContext.record Target record
     * @param {HTMLElement} event.editorContext.cell Target cell
     * @param {any} event.editorContext.value Cell value
     */
    onBeforeFinishCellEdit? : ((event: { grid: Grid, editorContext: { editor: Editor, column: Column, record: Model, cell: HTMLElement, value: any } }) => void)|string
    /**
     * Triggered before a widget is hidden. Return `false` to prevent the action.
     * @param {object} event Event object
     * @param {Widget} event.source The widget being hidden.
     */
    onBeforeHide? : ((event: { source: Widget }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires before the component stores a record's histogram data into the cache.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-beforeHistogramDataCacheSet)
     * @param {object} event Event object
     * @param {TimelineHistogram} event.source The component instance
     * @param {Model} event.record Record the histogram data of which is ready.
     * @param {object} event.data The record histogram data.
     */
    onBeforeHistogramDataCacheSet? : ((event: { source: TimelineHistogram, record: Model, data: object }) => void)|string
    /**
     * Fires on the owning Scheduler or Gantt widget before pan starts. Return `false` to prevent the operation.
     * @param {object} event Event object
     * @param {Event} event.event The native browser DOM event
     */
    onBeforePan? : ((event: { event: Event }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Grid before a paste action is performed, return `false` to prevent the action
     * @param {object} event Event object
     * @param {Grid} event.source Owner grid
     * @param {string} event.clipboardData The clipboardData about to be pasted
     * @param {GridLocation} event.targetCell The cell from which the paste will be started
     * @param {string} event.entityName 'cell' to distinguish this event from other beforePaste events
     */
    onBeforePaste? : ((event: { source: Grid, clipboardData: string, targetCell: GridLocation, entityName: string }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Grid before export started. Return `false` to cancel the export.
     * @param {object} event Event object
     * @param {object} event.config Export config
     */
    onBeforePdfExport? : ((event: { config: object }) => Promise<boolean>|boolean|void)|string
    /**
     * Fired before the [viewPreset](https://bryntum.com/products/gantt/docs/api/Scheduler/view/mixin/TimelineViewPresets#config-viewPreset) is changed.
     * @param {object} event Event object
     * @param {Scheduler} event.source This Scheduler instance.
     * @param {Date} event.startDate The new start date of the timeline.
     * @param {Date} event.endDate The new end date of the timeline.
     * @param {ViewPreset} event.from The outgoing ViewPreset.
     * @param {ViewPreset} event.to The ViewPreset being switched to.
     */
    onBeforePresetChange? : ((event: { source: Scheduler, startDate: Date, endDate: Date, from: ViewPreset, to: ViewPreset }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires before the component renders a row.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-beforeRenderHistogramRow)
     * @param {object} event Event object
     * @param {TimelineHistogram} event.source The component instance
     * @param {Model} event.record Record the histogram data of which is ready.
     * @param {HistogramConfig} event.histogramConfig Configuration object that will be applied to `histogramWidget`.
     * @param {Histogram} event.histogramWidget The underlying widget that is used to render a chart.
     */
    onBeforeRenderHistogramRow? : ((event: { source: TimelineHistogram, record: Model, histogramConfig: HistogramConfig, histogramWidget: Histogram }) => void)|string
    /**
     * Fires before the component renders a histogram in a cell.
     * @param {object} event Event object
     * @param {TimelineHistogram} event.source The component instance
     * @param {Model} event.record Record the histogram data of which is ready.
     * @param {HistogramConfig} event.histogramConfig Configuration object that will be applied to `histogramWidget`.
     * @param {Histogram} event.histogramWidget The underlying widget that is used to render a chart.
     */
    onBeforeRenderRecordHistogram? : ((event: { source: TimelineHistogram, record: Model, histogramConfig: HistogramConfig, histogramWidget: Histogram }) => void)|string
    /**
     * Fires before a row is rendered.
     * @param {object} event Event object
     * @param {Grid} event.source The firing Grid instance.
     * @param {Row} event.row The row about to be rendered.
     * @param {Model} event.record The record for the row.
     * @param {number} event.recordIndex The zero-based index of the record.
     */
    onBeforeRenderRow? : ((event: { source: Grid, row: Row, record: Model, recordIndex: number }) => void)|string
    /**
     * Grid rows are about to be rendered
     * @param {object} event Event object
     * @param {Grid} event.source This grid.
     */
    onBeforeRenderRows? : ((event: { source: Grid }) => void)|string
    /**
     * This event fires before row collapse is started.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-beforeRowCollapse)
     * @param {object} event Event object
     * @param {Model} event.record Record
     */
    onBeforeRowCollapse? : ((event: { record: Model }) => Promise<boolean>|boolean|void)|string
    /**
     * This event fires before row expand is started.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-beforeRowExpand)
     * @param {object} event Event object
     * @param {Model} event.record Record
     */
    onBeforeRowExpand? : ((event: { record: Model }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires before the selection changes. Returning `false` from a listener prevents the change
     * @param {object} event Event object
     * @param {string} event.action `'select'`/`'deselect'`
     * @param {'row'|'cell'} event.mode `'row'`/`'cell'`
     * @param {Grid} event.source
     * @param {Model[]} event.deselected The records to be deselected in this operation.
     * @param {Model[]} event.selected The records to be selected in this operation.
     * @param {Model[]} event.selection The records in the current selection, before applying `selected` and `deselected`
     * @param {GridLocation[]} event.deselectedCells The cells to be deselected in this operation.
     * @param {GridLocation[]} event.selectedCells The cells to be selected in this operation.
     * @param {GridLocation[]} event.cellSelection The cells in the current selection, before applying `selectedCells` and `deselectedCells`
     */
    onBeforeSelectionChange? : ((event: { action: string, mode: 'row'|'cell', source: Grid, deselected: Model[], selected: Model[], selection: Model[], deselectedCells: GridLocation[], selectedCells: GridLocation[], cellSelection: GridLocation[] }) => Promise<boolean>|boolean|void)|string
    /**
     * Fired before this container will load record values into its child fields. This is useful if you
     * want to modify the UI before data is loaded (e.g. set some input field to be readonly)
     * @param {object} event Event object
     * @param {Container} event.source The container
     * @param {Model} event.record The record
     */
    onBeforeSetRecord? : ((event: { source: Container, record: Model }) => void)|string
    /**
     * Triggered before a widget is shown. Return `false` to prevent the action.
     * @param {object} event Event object
     * @param {Widget|any} event.source The widget being shown
     */
    onBeforeShow? : ((event: { source: Widget|any }) => Promise<boolean>|boolean|void)|string
    /**
     * Fired when a group is going to be expanded or collapsed using the UI.
     * Returning `false` from a listener prevents the operation
     * @param {object} event Event object
     * @param {Model} event.groupRecord Group record
     * @param {boolean} event.collapse Collapsed (true) or expanded (false)
     * @param {Event} event.domEvent The user interaction event (eg a `click` event) if the toggle request was instigated by user interaction.
     */
    onBeforeToggleGroup? : ((event: { groupRecord: Model, collapse: boolean, domEvent: Event }) => Promise<boolean>|boolean|void)|string
    /**
     * Fired before a parent node record toggles its collapsed state.
     * @param {object} event Event object
     * @param {Grid} event.source The firing Grid instance.
     * @param {Model} event.record The record being toggled.
     * @param {boolean} event.collapse `true` if the node is being collapsed.
     */
    onBeforeToggleNode? : ((event: { source: Grid, record: Model, collapse: boolean }) => void)|string
    /**
     * Fires on the owning Grid when editing is cancelled
     * @param {object} event Event object
     * @param {Grid} event.source Owner grid
     * @param {GridLocation} event.editorContext Editing context
     * @param {Event} event.event Included if the cancellation was triggered by a DOM event
     */
    onCancelCellEdit? : ((event: { source: Grid, editorContext: GridLocation, event: Event }) => void)|string
    /**
     * Fires when any other event is fired from the object.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-catchAll)
     * @param {object} event Event object
     * @param {{[key: string]: any, type: string}} event.event The Object that contains event details
     * @param {string} event.event.type The type of the event which is caught by the listener
     */
    onCatchAll? : ((event: {[key: string]: any, type: string}) => void)|string
    /**
     * Fired when user clicks in a grid cell
     * @param {object} event Event object
     * @param {Grid} event.grid The grid instance
     * @param {Model} event.record The record representing the row
     * @param {Column} event.column The column to which the cell belongs
     * @param {HTMLElement} event.cellElement The cell HTML element
     * @param {HTMLElement} event.target The target element
     * @param {MouseEvent} event.event The native DOM event
     */
    onCellClick? : ((event: { grid: Grid, record: Model, column: Column, cellElement: HTMLElement, target: HTMLElement, event: MouseEvent }) => void)|string
    /**
     * Fired when user activates contextmenu in a grid cell
     * @param {object} event Event object
     * @param {Grid} event.grid The grid instance
     * @param {Model} event.record The record representing the row
     * @param {Column} event.column The column to which the cell belongs
     * @param {HTMLElement} event.cellElement The cell HTML element
     * @param {HTMLElement} event.target The target element
     * @param {MouseEvent} event.event The native DOM event
     */
    onCellContextMenu? : ((event: { grid: Grid, record: Model, column: Column, cellElement: HTMLElement, target: HTMLElement, event: MouseEvent }) => void)|string
    /**
     * Fired when user double clicks a grid cell
     * @param {object} event Event object
     * @param {Grid} event.grid The grid instance
     * @param {Model} event.record The record representing the row
     * @param {Column} event.column The column to which the cell belongs
     * @param {HTMLElement} event.cellElement The cell HTML element
     * @param {HTMLElement} event.target The target element
     * @param {MouseEvent} event.event The native DOM event
     */
    onCellDblClick? : ((event: { grid: Grid, record: Model, column: Column, cellElement: HTMLElement, target: HTMLElement, event: MouseEvent }) => void)|string
    /**
     * This event fires on the owning grid before the context menu is shown for a cell.
     * Allows manipulation of the items to show in the same way as in the [processItems](https://bryntum.com/products/gantt/docs/api/Grid/feature/CellMenu#config-processItems).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-cellMenuBeforeShow)
     * @param {object} event Event object
     * @param {Grid} event.source The grid
     * @param {Menu} event.menu The menu
     * @param {Record<string, MenuItemConfig>} event.items Menu item configs
     * @param {Column} event.column Column
     * @param {Model} event.record Record
     */
    onCellMenuBeforeShow? : ((event: { source: Grid, menu: Menu, items: Record<string, MenuItemConfig>, column: Column, record: Model }) => Promise<boolean>|boolean|void)|string
    /**
     * This event fires on the owning grid when an item is selected in the cell context menu.
     * @param {object} event Event object
     * @param {Grid} event.source The grid
     * @param {Menu} event.menu The menu
     * @param {MenuItem} event.item Selected menu item
     * @param {Column} event.column Column
     * @param {Model} event.record Record
     */
    onCellMenuItem? : ((event: { source: Grid, menu: Menu, item: MenuItem, column: Column, record: Model }) => void)|string
    /**
     * This event fires on the owning grid after the context menu is shown for a cell.
     * @param {object} event Event object
     * @param {Grid} event.source The grid
     * @param {Menu} event.menu The menu
     * @param {Record<string, MenuItemConfig>} event.items Menu item configs
     * @param {Column} event.column Column
     * @param {Model} event.record Record
     */
    onCellMenuShow? : ((event: { source: Grid, menu: Menu, items: Record<string, MenuItemConfig>, column: Column, record: Model }) => void)|string
    /**
     * This event fires on the owning grid when a check item is toggled in the cell context menu.
     * @param {object} event Event object
     * @param {Grid} event.source The grid
     * @param {Menu} event.menu The menu
     * @param {MenuItem} event.item Selected menu item
     * @param {Column} event.column Column
     * @param {Model} event.record Record
     * @param {boolean} event.checked Checked or not
     */
    onCellMenuToggleItem? : ((event: { source: Grid, menu: Menu, item: MenuItem, column: Column, record: Model, checked: boolean }) => void)|string
    /**
     * Fired when the mouse enters a cell
     * @param {object} event Event object
     * @param {Grid} event.source The grid instance
     * @param {Model} event.record The record representing the hovered cell
     * @param {Column} event.column The column currently hovered
     * @param {HTMLElement} event.cellElement The cell HTML element
     * @param {MouseEvent} event.event The native DOM event
     */
    onCellMouseEnter? : ((event: { source: Grid, record: Model, column: Column, cellElement: HTMLElement, event: MouseEvent }) => void)|string
    /**
     * Fired when the mouse leaves a cell
     * @param {object} event Event object
     * @param {Grid} event.source The grid instance
     * @param {Model} event.record The record representing the cell that the mouse left
     * @param {Column} event.column The column representing the cell that the mouse left
     * @param {HTMLElement} event.cellElement The cell HTML element that the mouse left
     * @param {MouseEvent} event.event The native DOM event
     */
    onCellMouseLeave? : ((event: { source: Grid, record: Model, column: Column, cellElement: HTMLElement, event: MouseEvent }) => void)|string
    /**
     * Fired when a user moves the mouse out of a grid cell
     * @param {object} event Event object
     * @param {Grid} event.grid The grid instance
     * @param {Model} event.record The record representing the row
     * @param {Column} event.column The column to which the cell belongs
     * @param {HTMLElement} event.cellElement The cell HTML element
     * @param {HTMLElement} event.target The target element
     * @param {MouseEvent} event.event The native DOM event
     */
    onCellMouseOut? : ((event: { grid: Grid, record: Model, column: Column, cellElement: HTMLElement, target: HTMLElement, event: MouseEvent }) => void)|string
    /**
     * Fired when user moves the mouse over a grid cell
     * @param {object} event Event object
     * @param {Grid} event.grid The grid instance
     * @param {Model} event.record The record representing the row
     * @param {Column} event.column The column to which the cell belongs
     * @param {HTMLElement} event.cellElement The cell HTML element
     * @param {HTMLElement} event.target The target element
     * @param {MouseEvent} event.event The native DOM event
     */
    onCellMouseOver? : ((event: { grid: Grid, record: Model, column: Column, cellElement: HTMLElement, target: HTMLElement, event: MouseEvent }) => void)|string
    /**
     * Fires when a Panel is collapsed using the [collapsible](https://bryntum.com/products/gantt/docs/api/Core/widget/Panel#config-collapsible) setting.
     * @param {object} event Event object
     * @param {Panel} event.source This Panel.
     */
    onCollapse? : ((event: { source: Panel }) => void)|string
    /**
     * Fired before a parent node record is collapsed.
     * @param {object} event Event object
     * @param {Grid} event.source The firing Grid instance.
     * @param {Model} event.record The record which has been collapsed.
     */
    onCollapseNode? : ((event: { source: Grid, record: Model }) => void)|string
    /**
     * This event is fired when a column is being dragged, and you can set the `valid` flag on t
     * @param {object} event Event object
     * @param {Grid} event.source The grid instance.
     * @param {Column} event.column The dragged column.
     * @param {Column} event.insertBefore The column before which the dragged column will be inserted.
     * @param {Event} event.event The browser event.
     * @param {object} event.context
     * @param {boolean} event.context.valid Set this to true or false to indicate whether the drop position is valid.
     */
    onColumnDrag? : ((event: { source: Grid, column: Column, insertBefore: Column, event: Event, context: { valid: boolean } }) => void)|string
    /**
     * This event is fired when a column drag gesture has started.
     * @param {object} event Event object
     * @param {Grid} event.source The grid instance.
     * @param {Column} event.column The dragged column.
     * @param {Event} event.event The browser event.
     */
    onColumnDragStart? : ((event: { source: Grid, column: Column, event: Event }) => void)|string
    /**
     * This event is always fired after a column is dropped. The `valid` param is true if the operation was not
     * vetoed and the column was moved in the column store.
     * @param {object} event Event object
     * @param {Grid} event.source The grid instance.
     * @param {Column} event.column The dragged column.
     * @param {Column} event.insertBefore The column before which the dragged column will be inserted.
     * @param {Column} event.newParent The new parent column.
     * @param {boolean} event.valid true if the operation was not vetoed.
     * @param {Event} event.event The browser event.
     * @param {string} event.region The region where the column was dropped.
     */
    onColumnDrop? : ((event: { source: Grid, column: Column, insertBefore: Column, newParent: Column, valid: boolean, event: Event, region: string }) => Promise<boolean>|boolean|void)|string
    /**
     * This event is fired after a resize gesture is completed.
     * @param {object} event Event object
     * @param {Grid} event.source The grid instance
     * @param {Column} event.column The resized column
     * @param {Event} event.domEvent The browser event
     */
    onColumnResize? : ((event: { source: Grid, column: Column, domEvent: Event }) => void)|string
    /**
     * This event is fired when a column resize gesture starts.
     * @param {object} event Event object
     * @param {Grid} event.source The grid instance
     * @param {Column} event.column The column
     * @param {Event} event.domEvent The browser event
     */
    onColumnResizeStart? : ((event: { source: Grid, column: Column, domEvent: Event }) => void)|string
    /**
     * This event fires on the owning widget when an item is selected in the context menu.
     * @param {object} event Event object
     * @param {Widget} event.source The owning widget
     * @param {Menu} event.menu The menu
     * @param {MenuItem} event.item Selected menu item
     */
    onContextMenuItem? : ((event: { source: Widget, menu: Menu, item: MenuItem }) => void)|string
    /**
     * This event fires on the owning widget when a check item is toggled in the context menu.
     * @param {object} event Event object
     * @param {Widget} event.source The owning widget
     * @param {Menu} event.menu The menu
     * @param {MenuItem} event.item Selected menu item
     * @param {boolean} event.checked Checked or not
     */
    onContextMenuToggleItem? : ((event: { source: Widget, menu: Menu, item: MenuItem, checked: boolean }) => void)|string
    /**
     * Fires on the owning Grid after a copy action is performed.
     * @param {object} event Event object
     * @param {Grid} event.source Owner grid
     * @param {GridLocation[]} event.cells The cells about to be copied or cut
     * @param {string} event.copiedDataString The concatenated data string that was copied or cut
     * @param {boolean} event.isCut `true` if this was a cut action
     * @param {string} event.entityName 'cell' to distinguish this event from other copy events
     */
    onCopy? : ((event: { source: Grid, cells: GridLocation[], copiedDataString: string, isCut: boolean, entityName: string }) => void)|string
    /**
     * Fires on the owning Scheduler/Gantt when the line indicating the current time is updated (see [currentTimeLineUpdateInterval](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/TimeRanges#config-currentTimeLineUpdateInterval)).
     * @param {object} event Event object
     * @param {Scheduler} event.source The scheduler
     * @param {Date} event.date The current date
     */
    onCurrentTimelineUpdate? : ((event: { source: Scheduler, date: Date }) => void)|string
    /**
     * Fired when data in the store changes.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-dataChange)
     * @param {object} event Event object
     * @param {Grid} event.source Owning grid
     * @param {Store} event.store The originating store
     * @param {'remove'|'removeAll'|'add'|'clearchanges'|'filter'|'update'|'dataset'|'replace'} event.action Name of action which triggered the change. May be one of:  * `'remove'` * `'removeAll'` * `'add'` * `'clearchanges'` * `'filter'` * `'update'` * `'dataset'` * `'replace'`
     * @param {Model} event.record Changed record, for actions that affects exactly one record (`'update'`)
     * @param {Model[]} event.records Changed records, passed for all actions except `'removeAll'`
     * @param {object} event.changes Passed for the `'update'` action, info on which record fields changed
     */
    onDataChange? : ((event: { source: Grid, store: Store, action: 'remove'|'removeAll'|'add'|'clearchanges'|'filter'|'update'|'dataset'|'replace', record: Model, records: Model[], changes: object }) => void)|string
    /**
     * Fired when the range of dates encapsulated by the UI changes. This will be when
     * moving a view in time by reconfiguring its [timeAxis](https://bryntum.com/products/gantt/docs/api/Scheduler/view/TimelineBase#config-timeAxis). This will happen
     * when zooming, or changing [viewPreset](https://bryntum.com/products/gantt/docs/api/Scheduler/view/TimelineBase#config-viewPreset).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-dateRangeChange)
     * @param {object} event Event object
     * @param {TimelineBase} event.source This Scheduler/Gantt instance.
     * @param {object} event.old The old date range
     * @param {Date} event.old.startDate the old start date.
     * @param {Date} event.old.endDate the old end date.
     * @param {object} event.new The new date range
     * @param {Date} event.new.startDate the new start date.
     * @param {Date} event.new.endDate the new end date.
     */
    onDateRangeChange? : ((event: { source: TimelineBase, old: { startDate: Date, endDate: Date }, new: { startDate: Date, endDate: Date } }) => void)|string
    /**
     * Fires when an object is destroyed.
     * @param {object} event Event object
     * @param {Base} event.source The Object that is being destroyed.
     */
    onDestroy? : ((event: { source: Base }) => void)|string
    /**
     * Fires when a field is mutated and the state of the [hasChanges](https://bryntum.com/products/gantt/docs/api/Core/widget/Container#property-hasChanges) property changes
     * @param {object} event Event object
     * @param {Container} event.source The container.
     * @param {boolean} event.dirty The dirty state of the Container - `true` if there are any fields which have been changed since initial load.
     */
    onDirtyStateChange? : ((event: { source: Container, dirty: boolean }) => void)|string
    /**
     * Fires while drag selecting. UI will update with current range, but the cells will not be selected until
     * mouse up. This event can be listened for to perform actions while drag selecting.
     * @param {object} event Event object
     * @param {Grid} event.source
     * @param {GridLocationConfig[]|GridLocation[]} event.selectedCells The cells that is currently being dragged over
     * @param {Model[]} event.selectedRecords The records that is currently being dragged over
     */
    onDragSelecting? : ((event: { source: Grid, selectedCells: GridLocationConfig[]|GridLocation[], selectedRecords: Model[] }) => void)|string
    /**
     * Fires when a Panel is expanded using the [collapsible](https://bryntum.com/products/gantt/docs/api/Core/widget/Panel#config-collapsible) setting.
     * @param {object} event Event object
     * @param {Panel} event.source This Panel.
     */
    onExpand? : ((event: { source: Panel }) => void)|string
    /**
     * Fired after a parent node record is expanded.
     * @param {object} event Event object
     * @param {Grid} event.source The firing Grid instance.
     * @param {Model} event.record The record which has been expanded.
     */
    onExpandNode? : ((event: { source: Grid, record: Model }) => void)|string
    /**
     * Fired when a file is dropped on the widget element
     * @param {object} event Event object
     * @param {Grid} event.source The owning Grid instance
     * @param {DataTransferItem} event.file The dropped file descriptor
     * @param {DragEvent} event.domEvent The native DragEvent
     */
    onFileDrop? : ((event: { source: Grid, file: DataTransferItem, domEvent: DragEvent }) => void)|string
    /**
     * Fired before the FillHandle dragging is finalized and values are applied to cells, return `false` to prevent the
     * drag operation from applying data changes.
     * @param {object} event Event object
     * @param {GridLocation} event.from The from cell
     * @param {GridLocation} event.to The to cell
     * @param {MouseEvent} event.domEvent The raw DOM event
     */
    onFillHandleBeforeDragFinalize? : ((event: { from: GridLocation, to: GridLocation, domEvent: MouseEvent }) => Promise<boolean>|boolean|void)|string
    /**
     * Fired while dragging the FillHandle.
     * @param {object} event Event object
     * @param {GridLocation} event.from The from cell
     * @param {GridLocation} event.to The to cell
     * @param {MouseEvent} event.domEvent The raw DOM event
     */
    onFillHandleDrag? : ((event: { from: GridLocation, to: GridLocation, domEvent: MouseEvent }) => void)|string
    /**
     * Fired when a FillHandle drag operation is aborted.
     */
    onFillHandleDragAbort? : (() => void)|string
    /**
     * Fired after a FillHandle drag operation.
     * @param {object} event Event object
     * @param {GridLocation} event.from The from cell
     * @param {GridLocation} event.to The to cell
     * @param {MouseEvent} event.domEvent The raw DOM event
     */
    onFillHandleDragEnd? : ((event: { from: GridLocation, to: GridLocation, domEvent: MouseEvent }) => void)|string
    /**
     * Fired when dragging of the FillHandle starts.
     * @param {object} event Event object
     * @param {GridLocation} event.cell Information about the column / record
     * @param {MouseEvent} event.domEvent The raw DOM event
     */
    onFillHandleDragStart? : ((event: { cell: GridLocation, domEvent: MouseEvent }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Grid when cell editing is finished
     * @param {object} event Event object
     * @param {Grid} event.grid Target grid
     * @param {GridLocation} event.editorContext Editing context
     * @param {Editor} event.editorContext.editor The Editor being used. Will contain an `inputField` property which is the field being used to perform the editing.
     * @param {Column} event.editorContext.column Target column
     * @param {Model} event.editorContext.record Target record
     * @param {HTMLElement} event.editorContext.cell Target cell
     * @param {any} event.editorContext.value Cell value
     */
    onFinishCellEdit? : ((event: { grid: Grid, editorContext: { editor: Editor, column: Column, record: Model, cell: HTMLElement, value: any } }) => void)|string
    /**
     * Fired when focus enters this Widget.
     * @param {object} event Event object
     * @param {Widget} event.source This Widget
     * @param {HTMLElement} event.fromElement The element which lost focus.
     * @param {HTMLElement} event.toElement The element which gained focus.
     * @param {Widget} event.fromWidget The widget which lost focus.
     * @param {Widget} event.toWidget The widget which gained focus.
     * @param {boolean} event.backwards `true` if the `toElement` is before the `fromElement` in document order.
     */
    onFocusIn? : ((event: { source: Widget, fromElement: HTMLElement, toElement: HTMLElement, fromWidget: Widget, toWidget: Widget, backwards: boolean }) => void)|string
    /**
     * Fired when focus exits this Widget's ownership tree. This is different from a `blur` event.
     * focus moving from within this Widget's ownership tree, even if there are floating widgets
     * will not trigger this event. This is when focus exits this widget completely.
     * @param {object} event Event object
     * @param {Widget} event.source This Widget
     * @param {HTMLElement} event.fromElement The element which lost focus.
     * @param {HTMLElement} event.toElement The element which gained focus.
     * @param {Widget} event.fromWidget The widget which lost focus.
     * @param {Widget} event.toWidget The widget which gained focus.
     * @param {boolean} event.backwards `true` if the `toElement` is before the `fromElement` in document order.
     */
    onFocusOut? : ((event: { source: Widget, fromElement: HTMLElement, toElement: HTMLElement, fromWidget: Widget, toWidget: Widget, backwards: boolean }) => void)|string
    /**
     * Fired before dragging starts, return false to prevent the drag operation.
     * @param {object} event Event object
     * @param {GridBase} event.source
     * @param {object} event.context
     * @param {Model[]} event.context.records The dragged row records
     * @param {MouseEvent|TouchEvent} event.event
     */
    onGridRowBeforeDragStart? : ((event: { source: GridBase, context: { records: Model[] }, event: MouseEvent|TouchEvent }) => Promise<boolean>|boolean|void)|string
    /**
     * Fired before the row drop operation is finalized. You can return false to abort the drop operation, or a
     * Promise yielding `true` / `false` which allows for asynchronous abort (e.g. first show user a confirmation dialog).
     * @param {object} event Event object
     * @param {GridBase} event.source
     * @param {object} event.context
     * @param {boolean} event.context.valid Set this to true or false to indicate whether the drop position is valid
     * @param {Model} event.context.insertBefore The record to insert before (`null` if inserting at last position of a parent node)
     * @param {Model} event.context.parent The parent record of the current drop position (only applicable for trees)
     * @param {Model[]} event.context.records The dragged row records
     * @param {RecordPositionContext[]} event.context.oldPositionContext An array of objects with information about the previous tree position. Objects contain the `record`, and its original `parentIndex` and `parentId` values
     * @param {MouseEvent} event.event
     */
    onGridRowBeforeDropFinalize? : ((event: { source: GridBase, context: { valid: boolean, insertBefore: Model, parent: Model, records: Model[], oldPositionContext: RecordPositionContext[] }, event: MouseEvent }) => Promise<boolean>|boolean|void)|string
    /**
     * Fired while the row is being dragged, in the listener function you have access to `context.insertBefore` a grid /
     * tree record, and additionally `context.parent` (a TreeNode) for trees. You can signal that the drop position is
     * valid or invalid by setting `context.valid = false;`
     * @param {object} event Event object
     * @param {GridBase} event.source
     * @param {object} event.context
     * @param {boolean} event.context.valid Set this to true or false to indicate whether the drop position is valid.
     * @param {Model} event.context.insertBefore The record to insert before (`null` if inserting at last position of a parent node)
     * @param {Model} event.context.parent The parent record of the current drop position (only applicable for trees)
     * @param {Model[]} event.context.records The dragged row records
     * @param {MouseEvent} event.event
     */
    onGridRowDrag? : ((event: { source: GridBase, context: { valid: boolean, insertBefore: Model, parent: Model, records: Model[] }, event: MouseEvent }) => void)|string
    /**
     * Fired when a row drag operation is aborted
     * @param {object} event Event object
     * @param {GridBase} event.source
     * @param {object} event.context
     * @param {Model[]} event.context.records The dragged row records
     * @param {MouseEvent} event.event
     */
    onGridRowDragAbort? : ((event: { source: GridBase, context: { records: Model[] }, event: MouseEvent }) => void)|string
    /**
     * Fired when dragging starts.
     * @param {object} event Event object
     * @param {GridBase} event.source
     * @param {object} event.context
     * @param {Model[]} event.context.records The dragged row records
     * @param {MouseEvent|TouchEvent} event.event
     */
    onGridRowDragStart? : ((event: { source: GridBase, context: { records: Model[] }, event: MouseEvent|TouchEvent }) => void)|string
    /**
     * Fired after the row drop operation has completed, regardless of validity
     * @param {object} event Event object
     * @param {GridBase} event.source
     * @param {object} event.context
     * @param {boolean} event.context.valid true or false depending on whether the drop position was valid
     * @param {Model} event.context.insertBefore The record to insert before (`null` if inserting at last position of a parent node)
     * @param {Model} event.context.parent The parent record of the current drop position (only applicable for trees)
     * @param {Model[]} event.context.records The dragged row records
     * @param {RecordPositionContext[]} event.context.oldPositionContext An array of objects with information about the previous tree position. Objects contain the record, and its original `parentIndex` and `parentId` values
     * @param {MouseEvent} event.event
     */
    onGridRowDrop? : ((event: { source: GridBase, context: { valid: boolean, insertBefore: Model, parent: Model, records: Model[], oldPositionContext: RecordPositionContext[] }, event: MouseEvent }) => void)|string
    /**
     * Fired when a grid header is clicked on.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-headerClick)
     * @param {object} event Event object
     * @param {Event} event.domEvent The triggering DOM event.
     * @param {Column} event.column The column clicked on.
     */
    onHeaderClick? : ((event: { domEvent: Event, column: Column }) => Promise<boolean>|boolean|void)|string
    /**
     * This event fires on the owning Grid before the context menu is shown for a header.
     * Allows manipulation of the items to show in the same way as in the [processItems](https://bryntum.com/products/gantt/docs/api/Grid/feature/HeaderMenu#config-processItems).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-headerMenuBeforeShow)
     * @param {object} event Event object
     * @param {Grid} event.source The grid
     * @param {Menu} event.menu The menu
     * @param {Record<string, MenuItemConfig>} event.items Menu item configs
     * @param {Column} event.column Column
     */
    onHeaderMenuBeforeShow? : ((event: { source: Grid, menu: Menu, items: Record<string, MenuItemConfig>, column: Column }) => Promise<boolean>|boolean|void)|string
    /**
     * This event fires on the owning Grid when an item is selected in the header context menu.
     * @param {object} event Event object
     * @param {Grid} event.source The grid
     * @param {Menu} event.menu The menu
     * @param {MenuItem} event.item Selected menu item
     * @param {Column} event.column Column
     */
    onHeaderMenuItem? : ((event: { source: Grid, menu: Menu, item: MenuItem, column: Column }) => void)|string
    /**
     * This event fires on the owning Grid after the context menu is shown for a header
     * @param {object} event Event object
     * @param {Grid} event.source The grid
     * @param {Menu} event.menu The menu
     * @param {Record<string, MenuItemConfig>} event.items Menu item configs
     * @param {Column} event.column Column
     */
    onHeaderMenuShow? : ((event: { source: Grid, menu: Menu, items: Record<string, MenuItemConfig>, column: Column }) => void)|string
    /**
     * This event fires on the owning Grid when a check item is toggled in the header context menu.
     * @param {object} event Event object
     * @param {Grid} event.source The grid
     * @param {Menu} event.menu The menu
     * @param {MenuItem} event.item Selected menu item
     * @param {Column} event.column Column
     * @param {boolean} event.checked Checked or not
     */
    onHeaderMenuToggleItem? : ((event: { source: Grid, menu: Menu, item: MenuItem, column: Column, checked: boolean }) => void)|string
    /**
     * Triggered after a widget was hidden
     * @param {object} event Event object
     * @param {Widget} event.source The widget
     */
    onHide? : ((event: { source: Widget }) => void)|string
    /**
     * Fires after the component retrieves a record's histogram data and stores
     * it into the cache.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-histogramDataCacheSet)
     * @param {object} event Event object
     * @param {TimelineHistogram} event.source The component instance
     * @param {Model} event.record Record the histogram data of which is ready.
     * @param {object} event.data The record histogram data.
     */
    onHistogramDataCacheSet? : ((event: { source: TimelineHistogram, record: Model, data: object }) => void)|string
    /**
     * Fires when row locking is enabled.
     * @param {object} event Event object
     * @param {GridBase} event.clone The created clone
     */
    onLockRows? : ((event: { clone: GridBase }) => void)|string
    /**
     * Mouse moved out from element in grid
     * @param {object} event Event object
     * @param {MouseEvent} event.event The native browser event
     */
    onMouseOut? : ((event: { event: MouseEvent }) => void)|string
    /**
     * Mouse moved in over element in grid
     * @param {object} event Event object
     * @param {MouseEvent} event.event The native browser event
     */
    onMouseOver? : ((event: { event: MouseEvent }) => void)|string
    /**
     * Triggered when a widget which had been in a non-visible state for any reason
     * achieves visibility.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-paint)
     * @param {object} event Event object
     * @param {Widget} event.source The widget being painted.
     * @param {boolean} event.firstPaint `true` if this is the first paint.
     */
    onPaint? : ((event: { source: Widget, firstPaint: boolean }) => void)|string
    /**
     * Fires on the owning Grid after a paste action is performed.
     * @param {object} event Event object
     * @param {Grid} event.source Owner grid
     * @param {string} event.clipboardData The clipboardData that was pasted
     * @param {Model[]} event.modifiedRecords The records which have been modified due to the paste action
     * @param {GridLocation} event.targetCell The cell from which the paste will be started
     * @param {string} event.entityName 'cell' to distinguish this event from other paste events
     */
    onPaste? : ((event: { source: Grid, clipboardData: string, modifiedRecords: Model[], targetCell: GridLocation, entityName: string }) => void)|string
    /**
     * Fires on the owning Grid when export has finished
     * @param {object} event Event object
     * @param {any} event.response Optional response, if received
     * @param {Error} event.error Optional error, if exception occurred
     */
    onPdfExport? : ((event: { response?: any, error?: Error }) => void)|string
    /**
     * Fired after the [viewPreset](https://bryntum.com/products/gantt/docs/api/Scheduler/view/mixin/TimelineViewPresets#config-viewPreset) has changed.
     * @param {object} event Event object
     * @param {Scheduler} event.source This Scheduler instance.
     * @param {Date} event.startDate The new start date of the timeline.
     * @param {Date} event.centerDate The new center date of the timeline.
     * @param {Date} event.endDate The new end date of the timeline.
     * @param {ViewPreset} event.from The outgoing ViewPreset.
     * @param {ViewPreset} event.to The ViewPreset being switched to.
     */
    onPresetChange? : ((event: { source: Scheduler, startDate: Date, centerDate: Date, endDate: Date, from: ViewPreset, to: ViewPreset }) => Promise<boolean>|boolean|void)|string
    /**
     * Fired when a Widget's read only state is toggled
     * @param {object} event Event object
     * @param {boolean} event.readOnly Read only or not
     */
    onReadOnly? : ((event: { readOnly: boolean }) => void)|string
    /**
     * This event is fired after a widget's elements have been synchronized due to a direct or indirect call
     * to [recompose](https://bryntum.com/products/gantt/docs/api/Core/widget/Widget#function-recompose), if this results in some change to the widget's rendered DOM elements.
     */
    onRecompose? : (() => void)|string
    /**
     * Fires after a row is rendered.
     * @param {object} event Event object
     * @param {Grid} event.source The firing Grid instance.
     * @param {Row} event.row The row that has been rendered.
     * @param {Model} event.record The record for the row.
     * @param {number} event.recordIndex The zero-based index of the record.
     */
    onRenderRow? : ((event: { source: Grid, row: Row, record: Model, recordIndex: number }) => void)|string
    /**
     * Grid rows have been rendered
     * @param {object} event Event object
     * @param {Grid} event.source This grid.
     */
    onRenderRows? : ((event: { source: Grid }) => void)|string
    /**
     * Fired when the encapsulating element of a Widget resizes *only when [monitorResize](https://bryntum.com/products/gantt/docs/api/Core/widget/Widget#config-monitorResize) is `true`*.
     * @param {object} event Event object
     * @param {Widget} event.source This Widget
     * @param {number} event.width The new width
     * @param {number} event.height The new height
     * @param {number} event.oldWidth The old width
     * @param {number} event.oldHeight The old height
     */
    onResize? : ((event: { source: Widget, width: number, height: number, oldWidth: number, oldHeight: number }) => void)|string
    /**
     * This event fires on the owning scheduler before the context menu is shown for a resource.
     * Allows manipulation of the items to show in the same way as in the [processItems](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/ResourceMenu#config-processItems).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-resourceMenuBeforeShow)
     * @param {object} event Event object
     * @param {Scheduler} event.source The grid
     * @param {Menu} event.menu The menu
     * @param {Record<string, MenuItemConfig>} event.items Menu item configs
     * @param {Column} event.column Column
     * @param {SchedulerResourceModel} event.resourceRecord Record
     */
    onResourceMenuBeforeShow? : ((event: { source: Scheduler, menu: Menu, items: Record<string, MenuItemConfig>, column: Column, resourceRecord: SchedulerResourceModel }) => Promise<boolean>|boolean|void)|string
    /**
     * This event fires on the owning scheduler when an item is selected in the context menu.
     * @param {object} event Event object
     * @param {Scheduler} event.source The grid
     * @param {Menu} event.menu The menu
     * @param {MenuItem} event.item Selected menu item
     * @param {SchedulerResourceModel} event.record Record
     */
    onResourceMenuItem? : ((event: { source: Scheduler, menu: Menu, item: MenuItem, record: SchedulerResourceModel }) => void)|string
    /**
     * This event fires on the owning scheduler after the context menu is shown for a resource.
     * @param {object} event Event object
     * @param {Scheduler} event.source The grid
     * @param {Menu} event.menu The menu
     * @param {Record<string, MenuItemConfig>} event.items Menu item configs
     * @param {SchedulerResourceModel} event.record Record
     */
    onResourceMenuShow? : ((event: { source: Scheduler, menu: Menu, items: Record<string, MenuItemConfig>, record: SchedulerResourceModel }) => void)|string
    /**
     * This event fires on the owning grid when a check item is toggled in the context menu.
     * @param {object} event Event object
     * @param {Scheduler} event.source The grid
     * @param {Menu} event.menu The menu
     * @param {MenuItem} event.item Selected menu item
     * @param {SchedulerResourceModel} event.resourceRecord Record
     * @param {boolean} event.checked Checked or not
     */
    onResourceMenuToggleItem? : ((event: { source: Scheduler, menu: Menu, item: MenuItem, resourceRecord: SchedulerResourceModel, checked: boolean }) => void)|string
    /**
     * Triggered for click on a resource nonworking time range. Only triggered if the ResourceNonWorkingTime feature is configured with
     * `enableMouseEvents: true`.
     * @param {object} event Event object
     * @param {SchedulerPro} event.source This Scheduler
     * @param {ResourceNonWorkingTime} event.feature The ResourceNonWorkingTime feature
     * @param {ResourceTimeRangeModel} event.resourceTimeRangeRecord Resource time range record
     * @param {SchedulerResourceModel} event.resourceRecord Resource record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onResourceNonWorkingTimeClick? : ((event: { source: SchedulerPro, feature: ResourceNonWorkingTime, resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, domEvent: MouseEvent }) => void)|string
    /**
     * Triggered for right-click on a resource nonworking time range. Only triggered if the ResourceNonWorkingTime feature is configured
     * with `enableMouseEvents: true`.
     * @param {object} event Event object
     * @param {SchedulerPro} event.source This Scheduler
     * @param {ResourceNonWorkingTime} event.feature The ResourceNonWorkingTime feature
     * @param {ResourceTimeRangeModel} event.resourceTimeRangeRecord Resource time range record
     * @param {SchedulerResourceModel} event.resourceRecord Resource record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onResourceNonWorkingTimeContextMenu? : ((event: { source: SchedulerPro, feature: ResourceNonWorkingTime, resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, domEvent: MouseEvent }) => void)|string
    /**
     * Triggered for double-click on a resource nonworking time range. Only triggered if the ResourceNonWorkingTime feature is configured
     * with `enableMouseEvents: true`.
     * @param {object} event Event object
     * @param {SchedulerPro} event.source This Scheduler
     * @param {ResourceNonWorkingTime} event.feature The ResourceNonWorkingTime feature
     * @param {ResourceTimeRangeModel} event.resourceTimeRangeRecord Resource time range record
     * @param {SchedulerResourceModel} event.resourceRecord Resource record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onResourceNonWorkingTimeDblClick? : ((event: { source: SchedulerPro, feature: ResourceNonWorkingTime, resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, domEvent: MouseEvent }) => void)|string
    /**
     * Triggered for mouse down ona resource nonworking time range. Only triggered if the ResourceNonWorkingTime feature is configured
     * with `enableMouseEvents: true`.
     * @param {object} event Event object
     * @param {SchedulerPro} event.source This Scheduler
     * @param {ResourceNonWorkingTime} event.feature The ResourceNonWorkingTime feature
     * @param {ResourceTimeRangeModel} event.resourceTimeRangeRecord Resource time range record
     * @param {SchedulerResourceModel} event.resourceRecord Resource record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onResourceNonWorkingTimeMouseDown? : ((event: { source: SchedulerPro, feature: ResourceNonWorkingTime, resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, domEvent: MouseEvent }) => void)|string
    /**
     * Triggered for mouse out of a resource nonworking time range. Only triggered if the ResourceNonWorkingTime feature is configured
     * with `enableMouseEvents: true`.
     * @param {object} event Event object
     * @param {SchedulerPro} event.source This Scheduler
     * @param {ResourceNonWorkingTime} event.feature The ResourceNonWorkingTime feature
     * @param {ResourceTimeRangeModel} event.resourceTimeRangeRecord Resource time range record
     * @param {SchedulerResourceModel} event.resourceRecord Resource record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onResourceNonWorkingTimeMouseOut? : ((event: { source: SchedulerPro, feature: ResourceNonWorkingTime, resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, domEvent: MouseEvent }) => void)|string
    /**
     * Triggered for mouse over on a resource nonworking time range. Only triggered if the ResourceNonWorkingTime feature is configured
     * with `enableMouseEvents: true`.
     * @param {object} event Event object
     * @param {SchedulerPro} event.source This Scheduler
     * @param {ResourceNonWorkingTime} event.feature The ResourceNonWorkingTime feature
     * @param {ResourceTimeRangeModel} event.resourceTimeRangeRecord Resource time range record
     * @param {SchedulerResourceModel} event.resourceRecord Resource record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onResourceNonWorkingTimeMouseOver? : ((event: { source: SchedulerPro, feature: ResourceNonWorkingTime, resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, domEvent: MouseEvent }) => void)|string
    /**
     * Triggered for mouse up ona resource nonworking time range. Only triggered if the ResourceNonWorkingTime feature is configured
     * with `enableMouseEvents: true`.
     * @param {object} event Event object
     * @param {SchedulerPro} event.source This Scheduler
     * @param {ResourceNonWorkingTime} event.feature The ResourceNonWorkingTime feature
     * @param {ResourceTimeRangeModel} event.resourceTimeRangeRecord Resource time range record
     * @param {SchedulerResourceModel} event.resourceRecord Resource record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onResourceNonWorkingTimeMouseUp? : ((event: { source: SchedulerPro, feature: ResourceNonWorkingTime, resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, domEvent: MouseEvent }) => void)|string
    /**
     * Triggered for click on a resource time range. Only triggered if the ResourceTimeRange feature is configured with
     * `enableMouseEvents: true`.
     * @param {object} event Event object
     * @param {Scheduler} event.source This Scheduler
     * @param {ResourceTimeRanges} event.feature The ResourceTimeRange feature
     * @param {ResourceTimeRangeModel} event.resourceTimeRangeRecord Resource time range record
     * @param {SchedulerResourceModel} event.resourceRecord Resource record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onResourceTimeRangeClick? : ((event: { source: Scheduler, feature: ResourceTimeRanges, resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, domEvent: MouseEvent }) => void)|string
    /**
     * Triggered for right-click on a resource time range. Only triggered if the ResourceTimeRange feature is configured
     * with `enableMouseEvents: true`.
     * @param {object} event Event object
     * @param {Scheduler} event.source This Scheduler
     * @param {ResourceTimeRanges} event.feature The ResourceTimeRange feature
     * @param {ResourceTimeRangeModel} event.resourceTimeRangeRecord Resource time range record
     * @param {SchedulerResourceModel} event.resourceRecord Resource record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onResourceTimeRangeContextMenu? : ((event: { source: Scheduler, feature: ResourceTimeRanges, resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, domEvent: MouseEvent }) => void)|string
    /**
     * Triggered for double-click on a resource time range. Only triggered if the ResourceTimeRange feature is configured
     * with `enableMouseEvents: true`.
     * @param {object} event Event object
     * @param {Scheduler} event.source This Scheduler
     * @param {ResourceTimeRanges} event.feature The ResourceTimeRange feature
     * @param {ResourceTimeRangeModel} event.resourceTimeRangeRecord Resource time range record
     * @param {SchedulerResourceModel} event.resourceRecord Resource record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onResourceTimeRangeDblClick? : ((event: { source: Scheduler, feature: ResourceTimeRanges, resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, domEvent: MouseEvent }) => void)|string
    /**
     * Triggered for mouse down ona resource time range. Only triggered if the ResourceTimeRange feature is configured
     * with `enableMouseEvents: true`.
     * @param {object} event Event object
     * @param {Scheduler} event.source This Scheduler
     * @param {ResourceTimeRanges} event.feature The ResourceTimeRange feature
     * @param {ResourceTimeRangeModel} event.resourceTimeRangeRecord Resource time range record
     * @param {SchedulerResourceModel} event.resourceRecord Resource record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onResourceTimeRangeMouseDown? : ((event: { source: Scheduler, feature: ResourceTimeRanges, resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, domEvent: MouseEvent }) => void)|string
    /**
     * Triggered for mouse out of a resource time range. Only triggered if the ResourceTimeRange feature is configured
     * with `enableMouseEvents: true`.
     * @param {object} event Event object
     * @param {Scheduler} event.source This Scheduler
     * @param {ResourceTimeRanges} event.feature The ResourceTimeRange feature
     * @param {ResourceTimeRangeModel} event.resourceTimeRangeRecord Resource time range record
     * @param {SchedulerResourceModel} event.resourceRecord Resource record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onResourceTimeRangeMouseOut? : ((event: { source: Scheduler, feature: ResourceTimeRanges, resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, domEvent: MouseEvent }) => void)|string
    /**
     * Triggered for mouse over on a resource time range. Only triggered if the ResourceTimeRange feature is configured
     * with `enableMouseEvents: true`.
     * @param {object} event Event object
     * @param {Scheduler} event.source This Scheduler
     * @param {ResourceTimeRanges} event.feature The ResourceTimeRange feature
     * @param {ResourceTimeRangeModel} event.resourceTimeRangeRecord Resource time range record
     * @param {SchedulerResourceModel} event.resourceRecord Resource record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onResourceTimeRangeMouseOver? : ((event: { source: Scheduler, feature: ResourceTimeRanges, resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, domEvent: MouseEvent }) => void)|string
    /**
     * Triggered for mouse up ona resource time range. Only triggered if the ResourceTimeRange feature is configured
     * with `enableMouseEvents: true`.
     * @param {object} event Event object
     * @param {Scheduler} event.source This Scheduler
     * @param {ResourceTimeRanges} event.feature The ResourceTimeRange feature
     * @param {ResourceTimeRangeModel} event.resourceTimeRangeRecord Resource time range record
     * @param {SchedulerResourceModel} event.resourceRecord Resource record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onResourceTimeRangeMouseUp? : ((event: { source: Scheduler, feature: ResourceTimeRanges, resourceTimeRangeRecord: ResourceTimeRangeModel, resourceRecord: SchedulerResourceModel, domEvent: MouseEvent }) => void)|string
    /**
     * Grid resize lead to a new responsive level being applied
     * @param {object} event Event object
     * @param {Grid} event.grid Grid that was resized
     * @param {string} event.level New responsive level (small, large, etc)
     * @param {number} event.width New width in px
     * @param {string} event.oldLevel Old responsive level
     * @param {number} event.oldWidth Old width in px
     */
    onResponsive? : ((event: { grid: Grid, level: string, width: number, oldLevel: string, oldWidth: number }) => void)|string
    /**
     * This event fires when a row has finished collapsing.
     * @param {object} event Event object
     * @param {Model} event.record Record
     */
    onRowCollapse? : ((event: { record: Model }) => void)|string
    /**
     * This event fires when a row expand has finished expanding.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-rowExpand)
     * @param {object} event Event object
     * @param {Model} event.record Record
     * @param {object} event.expandedElements An object with the Grid region name as property and the expanded body element as value
     * @param {Widget} event.widget In case of expanding a Widget, this will be a reference to the instance created by the actual expansion. If there is multiple Grid regions, use the `widgets` param instead.
     * @param {object} event.widgets In case of expanding a Widget, this will be an object with the Grid region name as property and the reference to the widget instance created by the actual expansion
     */
    onRowExpand? : ((event: { record: Model, expandedElements: object, widget: Widget, widgets: object }) => void)|string
    /**
     * Fired when the mouse enters a row
     * @param {object} event Event object
     * @param {Grid} event.source The grid instance
     * @param {Model} event.record The record representing the hovered row
     * @param {Column} event.column The column currently hovered
     * @param {HTMLElement} event.cellElement The cell HTML element
     * @param {MouseEvent} event.event The native DOM event
     */
    onRowMouseEnter? : ((event: { source: Grid, record: Model, column: Column, cellElement: HTMLElement, event: MouseEvent }) => void)|string
    /**
     * Fired when the mouse leaves a row
     * @param {object} event Event object
     * @param {Grid} event.source The grid instance
     * @param {Model} event.record The record representing the row that the mouse left
     * @param {HTMLElement} event.cellElement The cell HTML element that the mouse left
     * @param {MouseEvent} event.event The native DOM event
     */
    onRowMouseLeave? : ((event: { source: Grid, record: Model, cellElement: HTMLElement, event: MouseEvent }) => void)|string
    /**
     * Grid has scrolled vertically
     * @param {object} event Event object
     * @param {Grid} event.source The firing Grid instance.
     * @param {number} event.scrollTop The vertical scroll position.
     */
    onScroll? : ((event: { source: Grid, scrollTop: number }) => void)|string
    /**
     * The selection has been changed.
     * @param {object} event Event object
     * @param {'select'|'deselect'} event.action `'select'`/`'deselect'`
     * @param {'row'|'cell'} event.mode `'row'`/`'cell'`
     * @param {Grid} event.source
     * @param {Model[]} event.deselected The records deselected in this operation.
     * @param {Model[]} event.selected The records selected in this operation.
     * @param {Model[]} event.selection The records in the new selection.
     * @param {GridLocation[]} event.deselectedCells The cells deselected in this operation.
     * @param {GridLocation[]} event.selectedCells The cells selected in this operation.
     * @param {GridLocation[]} event.cellSelection The cells in the new selection.
     */
    onSelectionChange? : ((event: { action: 'select'|'deselect', mode: 'row'|'cell', source: Grid, deselected: Model[], selected: Model[], selection: Model[], deselectedCells: GridLocation[], selectedCells: GridLocation[], cellSelection: GridLocation[] }) => void)|string
    /**
     * The selectionMode configuration has been changed.
     * @param {object} event Event object
     * @param {object} event.selectionMode The new [selectionMode](https://bryntum.com/products/gantt/docs/api/Grid/view/mixin/GridSelection#config-selectionMode)
     */
    onSelectionModeChange? : ((event: { selectionMode: object }) => void)|string
    /**
     * Triggered after a widget is shown.
     * @param {object} event Event object
     * @param {Widget} event.source The widget
     */
    onShow? : ((event: { source: Widget }) => void)|string
    /**
     * Fires when splitting the Grid.
     * @param {object} event Event object
     * @param {GridBase[]} event.subViews The sub views created by the split
     * @param {object} event.options The options passed to the split call
     * @param {'horizontal'|'vertical'|'both'} event.options.direction The direction of the split
     * @param {Column} event.options.atColumn The column to split at
     * @param {Model} event.options.atRecord The record to split at
     */
    onSplit? : ((event: { subViews: GridBase[], options: { direction: 'horizontal'|'vertical'|'both', atColumn: Column, atRecord: Model } }) => void)|string
    /**
     * Fired by the Grid when the collapse icon is clicked. Return `false` to prevent the default collapse action,
     * if you want to implement your own behavior.
     * @param {object} event Event object
     * @param {Grid} event.source The Grid instance.
     * @param {SubGrid} event.subGrid The subgrid
     * @param {Event} event.domEvent The native DOM event
     */
    onSplitterCollapseClick? : ((event: { source: Grid, subGrid: SubGrid, domEvent: Event }) => Promise<boolean>|boolean|void)|string
    /**
     * Fired by the Grid after a sub-grid has been resized using the splitter
     * @param {object} event Event object
     * @param {Grid} event.source The Grid instance.
     * @param {SubGrid} event.subGrid The resized subgrid
     * @param {Event} event.domEvent The native DOM event
     */
    onSplitterDragEnd? : ((event: { source: Grid, subGrid: SubGrid, domEvent: Event }) => void)|string
    /**
     * Fired by the Grid when a sub-grid resize gesture starts
     * @param {object} event Event object
     * @param {Grid} event.source The Grid instance.
     * @param {SubGrid} event.subGrid The subgrid about to be resized
     * @param {Event} event.domEvent The native DOM event
     */
    onSplitterDragStart? : ((event: { source: Grid, subGrid: SubGrid, domEvent: Event }) => void)|string
    /**
     * Fired by the Grid when the expand icon is clicked. Return `false` to prevent the default expand action,
     * if you want to implement your own behavior.
     * @param {object} event Event object
     * @param {Grid} event.source The Grid instance.
     * @param {SubGrid} event.subGrid The subgrid
     * @param {Event} event.domEvent The native DOM event
     */
    onSplitterExpandClick? : ((event: { source: Grid, subGrid: SubGrid, domEvent: Event }) => Promise<boolean>|boolean|void)|string
    /**
     * Fires on the owning Grid when editing starts
     * @param {object} event Event object
     * @param {Grid} event.source Owner grid
     * @param {GridLocation} event.editorContext Editing context
     * @param {Editor} event.editorContext.editor The Editor being used. Will contain an `inputField` property which is the field being used to perform the editing.
     * @param {Column} event.editorContext.column Target column
     * @param {Model} event.editorContext.record Target record
     * @param {HTMLElement} event.editorContext.cell Target cell
     * @param {any} event.editorContext.value Cell value
     */
    onStartCellEdit? : ((event: { source: Grid, editorContext: { editor: Editor, column: Column, record: Model, cell: HTMLElement, value: any } }) => void)|string
    /**
     * Fires after a sub grid is collapsed.
     * @param {object} event Event object
     * @param {Grid} event.source The firing Grid instance
     * @param {SubGrid} event.subGrid The sub grid instance
     */
    onSubGridCollapse? : ((event: { source: Grid, subGrid: SubGrid }) => void)|string
    /**
     * Fires after a sub grid is expanded.
     * @param {object} event Event object
     * @param {Grid} event.source The firing Grid instance
     * @param {SubGrid} event.subGrid The sub grid instance
     */
    onSubGridExpand? : ((event: { source: Grid, subGrid: SubGrid }) => void)|string
    /**
     * Fired when the tick size changes.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-tickSizeChange)
     * @param {object} event Event object
     * @param {Scheduler} event.source This Scheduler instance.
     * @param {number} event.tickSize The tick size in pixels
     */
    onTickSizeChange? : ((event: { source: Scheduler, tickSize: number }) => void)|string
    /**
     * Fired when the timeaxis has changed, for example by zooming or configuring a new time span.
     * @param {object} event Event object
     * @param {Scheduler} event.source This Scheduler
     * @param {object} event.config Config object used to reconfigure the time axis.
     * @param {Date} event.config.startDate New start date (if supplied)
     * @param {Date} event.config.endDate New end date (if supplied)
     */
    onTimeAxisChange? : ((event: { source: Scheduler, config: { startDate: Date, endDate: Date } }) => void)|string
    /**
     * Fires after a click on a time axis cell
     * @param {object} event Event object
     * @param {SchedulerTimeAxisColumn|VerticalTimeAxisColumn} event.source The column object
     * @param {Date} event.startDate The start date of the header cell
     * @param {Date} event.endDate The end date of the header cell
     * @param {Event} event.event The event object
     */
    onTimeAxisHeaderClick? : ((event: { source: SchedulerTimeAxisColumn|VerticalTimeAxisColumn, startDate: Date, endDate: Date, event: Event }) => void)|string
    /**
     * Fires after a right click on a time axis cell
     * @param {object} event Event object
     * @param {SchedulerTimeAxisColumn|VerticalTimeAxisColumn} event.source The column object
     * @param {Date} event.startDate The start date of the header cell
     * @param {Date} event.endDate The end date of the header cell
     * @param {Event} event.event The event object
     */
    onTimeAxisHeaderContextMenu? : ((event: { source: SchedulerTimeAxisColumn|VerticalTimeAxisColumn, startDate: Date, endDate: Date, event: Event }) => void)|string
    /**
     * Fires after a double click on a time axis cell
     * @param {object} event Event object
     * @param {SchedulerTimeAxisColumn|VerticalTimeAxisColumn} event.source The column object
     * @param {Date} event.startDate The start date of the header cell
     * @param {Date} event.endDate The end date of the header cell
     * @param {Event} event.event The event object
     */
    onTimeAxisHeaderDblClick? : ((event: { source: SchedulerTimeAxisColumn|VerticalTimeAxisColumn, startDate: Date, endDate: Date, event: Event }) => void)|string
    /**
     * This event fires on the owning Scheduler or Gantt widget before the context menu is shown for the time axis header.
     * Allows manipulation of the items to show in the same way as in the [processItems](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/TimeAxisHeaderMenu#config-processItems).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-timeAxisHeaderMenuBeforeShow)
     * @param {object} event Event object
     * @param {Scheduler} event.source The scheduler
     * @param {Menu} event.menu The menu
     * @param {Record<string, MenuItemConfig>} event.items Menu item configs
     * @param {Column} event.column Time axis column
     */
    onTimeAxisHeaderMenuBeforeShow? : ((event: { source: Scheduler, menu: Menu, items: Record<string, MenuItemConfig>, column: Column }) => Promise<boolean>|boolean|void)|string
    /**
     * This event fires on the owning Scheduler or Gantt widget when an item is selected in the header context menu.
     * @param {object} event Event object
     * @param {Scheduler} event.source The scheduler
     * @param {Menu} event.menu The menu
     * @param {MenuItem} event.item Selected menu item
     * @param {Column} event.column Time axis column
     */
    onTimeAxisHeaderMenuItem? : ((event: { source: Scheduler, menu: Menu, item: MenuItem, column: Column }) => void)|string
    /**
     * This event fires on the owning Scheduler or Gantt widget after the context menu is shown for a header
     * @param {object} event Event object
     * @param {Scheduler} event.source The scheduler
     * @param {Menu} event.menu The menu
     * @param {Record<string, MenuItemConfig>} event.items Menu item configs
     * @param {Column} event.column Time axis column
     */
    onTimeAxisHeaderMenuShow? : ((event: { source: Scheduler, menu: Menu, items: Record<string, MenuItemConfig>, column: Column }) => void)|string
    /**
     * Fired when the pointer-activated [timelineContext](https://bryntum.com/products/gantt/docs/api/Scheduler/view/mixin/TimelineDomEvents#property-timelineContext) has changed.
     * @param {object} event Event object
     * @param {TimelineContext} event.oldContext The tick/resource context being deactivated.
     * @param {TimelineContext} event.context The tick/resource context being activated.
     */
    onTimelineContextChange? : ((event: { oldContext: TimelineContext, context: TimelineContext }) => void)|string
    /**
     * Fired when the *scheduler* viewport (not the overall Scheduler element) changes size.
     * This happens when the grid changes height, or when the subgrid which encapsulates the
     * scheduler column changes width.
     * @param {object} event Event object
     * @param {Widget} event.source This Scheduler
     * @param {number} event.width The new width
     * @param {number} event.height The new height
     * @param {number} event.oldWidth The old width
     * @param {number} event.oldHeight The old height
     */
    onTimelineViewportResize? : ((event: { source: Widget, width: number, height: number, oldWidth: number, oldHeight: number }) => void)|string
    /**
     * Fired on the owning Scheduler or Gantt widget when a click happens on a time range header element
     * @param {object} event Event object
     * @param {Scheduler} event.source Scheduler instance
     * @param {TimeSpan} event.timeRangeRecord The record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onTimeRangeHeaderClick? : ((event: { source: Scheduler, timeRangeRecord: TimeSpan, domEvent: MouseEvent }) => void)|string
    /**
     * Fired on the owning Scheduler or Gantt widget when a right click happens on a time range header element
     * @param {object} event Event object
     * @param {Scheduler} event.source Scheduler instance
     * @param {TimeSpan} event.timeRangeRecord The record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onTimeRangeHeaderContextMenu? : ((event: { source: Scheduler, timeRangeRecord: TimeSpan, domEvent: MouseEvent }) => void)|string
    /**
     * Fired on the owning Scheduler or Gantt widget when a double click happens on a time range header element
     * @param {object} event Event object
     * @param {Scheduler} event.source Scheduler instance
     * @param {TimeSpan} event.timeRangeRecord The record
     * @param {MouseEvent} event.domEvent Browser event
     */
    onTimeRangeHeaderDblClick? : ((event: { source: Scheduler, timeRangeRecord: TimeSpan, domEvent: MouseEvent }) => void)|string
    /**
     * Group expanded or collapsed
     * @param {object} event Event object
     * @param {Model} event.groupRecord Group record
     * @param {boolean} event.collapse Collapsed (true) or expanded (false)
     * @param {boolean} event.allRecords True if this event is part of toggling all groups
     */
    onToggleGroup? : ((event: { groupRecord: Model, collapse: boolean, allRecords?: boolean }) => void)|string
    /**
     * Fired after a parent node record toggles its collapsed state.
     * @param {object} event Event object
     * @param {Model} event.record The record being toggled.
     * @param {boolean} event.collapse `true` if the node is being collapsed.
     */
    onToggleNode? : ((event: { record: Model, collapse: boolean }) => void)|string
    /**
     * A header [tool](https://bryntum.com/products/gantt/docs/api/Core/widget/Panel#config-tools) has been clicked.
     * @param {object} event Event object
     * @param {Tool} event.source This Panel.
     * @param {Tool} event.tool The tool which is being clicked.
     */
    onToolClick? : ((event: { source: Tool, tool: Tool }) => void)|string
    /**
     * Fires when row locking is disabled.
     * @param {object} event Event object
     * @param {GridBase} event.clone The locked clone that will be destroyed
     */
    onUnlockRows? : ((event: { clone: GridBase }) => void)|string
    /**
     * Fires when un-splitting the Grid.
     */
    onUnsplit? : (() => void)|string
    /**
     * Fired when the range of dates visible within the viewport changes. This will be when
     * scrolling along a time axis.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/view/ResourceUtilization#event-visibleDateRangeChange)
     * @param {object} event Event object
     * @param {Scheduler} event.source This Scheduler instance.
     * @param {object} event.old The old date range
     * @param {Date} event.old.startDate the old start date.
     * @param {Date} event.old.endDate the old end date.
     * @param {object} event.new The new date range
     * @param {Date} event.new.startDate the new start date.
     * @param {Date} event.new.endDate the new end date.
     */
    onVisibleDateRangeChange? : ((event: { source: Scheduler, old: { startDate: Date, endDate: Date }, new: { startDate: Date, endDate: Date } }) => void)|string

    // Features
    /**
     * Allows using `[Ctrl/CMD + C]`, `[Ctrl/CMD + X]` and `[Ctrl/CMD + V]` to cut, copy and paste cell or cell ranges. Also
     * makes cut, copy and paste actions available via the cell context menu.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/CellCopyPaste)
     */
    cellCopyPasteFeature? : object|boolean|string|CellCopyPaste|CellCopyPasteConfig
    /**
     * Extends the [CellEdit](https://bryntum.com/products/gantt/docs/api/Grid/feature/CellEdit) to encapsulate SchedulerPro functionality.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/feature/CellEdit)
     */
    cellEditFeature? : object|boolean|string|SchedulerProCellEdit|SchedulerProCellEditConfig
    /**
     * Right click to display context menu for cells.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/CellMenu)
     */
    cellMenuFeature? : object|boolean|string|CellMenu|CellMenuConfig
    /**
     * Displays a tooltip when hovering cells.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/CellTooltip)
     */
    cellTooltipFeature? : object|boolean|string|CellTooltip|CellTooltipConfig
    /**
     * Enables the [autoWidth](https://bryntum.com/products/gantt/docs/api/Grid/column/Column#config-autoWidth) config for a grid's columns.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/ColumnAutoWidth)
     */
    columnAutoWidthFeature? : object|boolean|string|ColumnAutoWidth|ColumnAutoWidthConfig
    /**
     * Displays a toolbar while dragging column headers. Drop on a button in the toolbar to activate a certain function,
     * for example to group by that column. This feature simplifies certain operations on touch devices.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/ColumnDragToolbar)
     */
    columnDragToolbarFeature? : object|boolean|string|ColumnDragToolbar|ColumnDragToolbarConfig
    /**
     * Displays column lines for ticks, with a different styling for major ticks (by default they are darker). If this
     * feature is disabled, no lines are shown. If it's enabled, line are shown for the tick level which is set in current
     * ViewPreset. Please see [columnLinesFor](https://bryntum.com/products/gantt/docs/api/Scheduler/preset/ViewPreset#field-columnLinesFor) config for details.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/ColumnLines)
     */
    columnLinesFeature? : object|boolean|string|ColumnLines|ColumnLinesConfig
    /**
     * Displays a column picker (to show/hide columns) in the header context menu. Columns can be displayed in sub menus
     * by region or tag. Grouped headers are displayed as menu hierarchies.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/ColumnPicker)
     */
    columnPickerFeature? : object|boolean|string|ColumnPicker|ColumnPickerConfig
    /**
     * Allows user to rename columns by either right-clicking column header or using keyboard shortcuts when column header
     * is focused.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/ColumnRename)
     */
    columnRenameFeature? : object|boolean|string|ColumnRename|ColumnRenameConfig
    /**
     * Allows user to reorder columns by dragging headers. To get notified about column reorder listen to `change` event
     * on [columns](https://bryntum.com/products/gantt/docs/api/Grid/data/ColumnStore) store.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/ColumnReorder)
     */
    columnReorderFeature? : object|boolean|string|ColumnReorder|ColumnReorderConfig
    /**
     * Enables user to resize columns by dragging a handle on the right hand side of the header. To get notified about column
     * resize listen to `change` event on [columns](https://bryntum.com/products/gantt/docs/api/Grid/data/ColumnStore) store.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/ColumnResize)
     */
    columnResizeFeature? : object|boolean|string|ColumnResize|ColumnResizeConfig
    /**
     * A feature that allows exporting Grid data to Excel or CSV without involving the server. It uses [TableExporter](https://bryntum.com/products/gantt/docs/api/Grid/util/TableExporter)
     * class as data provider, [zipcelx library](https://www.npmjs.com/package/zipcelx)
     * forked and adjusted to support [column width config](https://github.com/bryntum/zipcelx/tree/column-width-build)
     * and [Microsoft XML specification](https://msdn.microsoft.com/en-us/library/office/documentformat.openxml.spreadsheet.aspx).
     * Zipcelx should be either in global scope (window) or can be provided with [zipcelx](https://bryntum.com/products/gantt/docs/api/Grid/feature/experimental/ExcelExporter#config-zipcelx) config.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/experimental/ExcelExporter)
     */
    excelExporterFeature? : object|boolean|string|GridExcelExporter|GridExcelExporterConfig
    /**
     * An experimental feature that lets users drop files on a Widget. The widget fires an event when a file is dropped onto it.
     * In the event, you get access to the raw files as strings, that were parsed by calling `readAsBinaryString`.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/experimental/FileDrop)
     */
    fileDropFeature? : object|boolean|string|FileDrop|FileDropConfig
    /**
     * This feature adds a fill handle to a Grid range selection, which when dragged, fills the cells being dragged over
     * with values based on the values in the original selected range. This is similar to functionality normally seen in
     * various spreadsheet applications.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/FillHandle)
     */
    fillHandleFeature? : object|boolean|string|FillHandle|FillHandleConfig
    /**
     * Feature that allows filtering of the grid by settings filters on columns. The actual filtering is done by the store.
     * For info on programmatically handling filters, see [StoreFilter](https://bryntum.com/products/gantt/docs/api/Core/data/mixin/StoreFilter).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/Filter)
     */
    filterFeature? : object|boolean|string|Filter|FilterConfig
    /**
     * Feature that allows filtering of the grid by entering filters on column headers.
     * The actual filtering is done by the store.
     * For info on programmatically handling filters, see [StoreFilter](https://bryntum.com/products/gantt/docs/api/Core/data/mixin/StoreFilter).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/FilterBar)
     */
    filterBarFeature? : object|boolean|string|FilterBar|FilterBarConfig
    /**
     * Enables rendering and handling of row groups. The actual grouping is done in the store, but triggered by `[shift]` +
     * clicking headers or by using two finger tap (one on header, one anywhere on grid). Use `[shift]` + `[alt]` + click to
     * remove a column grouper.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/Group)
     */
    groupFeature? : object|boolean|string|Group|GroupConfig
    /**
     * A special version of the Grid GroupSummary feature that enables summaries within scheduler. To use a single summary
     * it is easiest to configure [renderer](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/GroupSummary#config-renderer), for multiple summaries see [summaries](#Scheduler/feature/GroupSummary#config-summaries).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/GroupSummary)
     */
    groupSummaryFeature? : object|boolean|string|GroupSummary|GroupSummaryConfig
    /**
     * Right click column header or focus it and press SPACE key to show the context menu for headers.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/HeaderMenu)
     */
    headerMenuFeature? : object|boolean|string|HeaderMenu|HeaderMenuConfig
    /**
     * Enables users to click and drag to zoom to a date range in Scheduler's header time axis. Only supported in horizontal
     * mode.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/HeaderZoom)
     */
    headerZoomFeature? : object|boolean|string|HeaderZoom|HeaderZoomConfig
    /**
     * This feature allows records which satisfy a certain condition to be locked at the top of the grid.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/LockRows)
     */
    lockRowsFeature? : object|boolean|string|GridLockRows|GridLockRowsConfig
    /**
     * This feature merges cells that have the same value in sorted (or [optionally](https://bryntum.com/products/gantt/docs/api/Grid/feature/MergeCells#config-sortedOnly) any) columns
     * configured to [mergeCells](https://bryntum.com/products/gantt/docs/api/Grid/column/Column#config-mergeCells).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/MergeCells)
     */
    mergeCellsFeature? : object|boolean|string|MergeCells|MergeCellsConfig
    /**
     * Feature that allows styling of weekends (and other non-working time) by adding timeRanges for those days.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/NonWorkingTime)
     */
    nonWorkingTimeFeature? : object|boolean|string|NonWorkingTime|NonWorkingTimeConfig
    /**
     * Makes the scheduler's timeline pannable by dragging with the mouse. Try it out in the demo below.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/Pan)
     */
    panFeature? : object|boolean|string|Pan|PanConfig
    /**
     * Generates PDF/PNG files from the Grid component.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/export/PdfExport)
     */
    pdfExportFeature? : object|boolean|string|GridPdfExport|GridPdfExportConfig
    /**
     * Allows printing Grid contents using browser print dialog.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/export/Print)
     */
    printFeature? : object|boolean|string|GridPrint|GridPrintConfig
    /**
     * Feature that allows the user to search in a column by focusing a cell and typing. Navigate between hits using the
     * keyboard, [f3] or [ctrl]/[cmd] + [g] moves to next, also pressing [shift] moves to previous.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/QuickFind)
     */
    quickFindFeature? : object|boolean|string|QuickFind|QuickFindConfig
    /**
     * Makes the splitter between grid [sections](https://bryntum.com/products/gantt/docs/api/Grid/view/SubGrid) draggable, to let users resize, and
     * collapse/expand the sections.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/RegionResize)
     */
    regionResizeFeature? : object|boolean|string|RegionResize|RegionResizeConfig
    /**
     * Applicable only for Scheduler in `vertical` mode. Right click resource header cells to display a context menu.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/ResourceMenu)
     */
    resourceMenuFeature? : object|boolean|string|ResourceMenu|ResourceMenuConfig
    /**
     * Feature that highlights the non-working intervals for resources based on their [calendar](https://bryntum.com/products/gantt/docs/api/SchedulerPro/model/ResourceModel#field-calendar).
     * If a resource has no calendar defined, the project's calendar will be used. The non-working time interval can
     * also be recurring. You can find a live example showing how to achieve this in the [Resource Non-Working Time Demo](../examples/resource-non-working-time/).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/feature/ResourceNonWorkingTime)
     */
    resourceNonWorkingTimeFeature? : object|boolean|string|ResourceNonWorkingTime|ResourceNonWorkingTimeConfig
    /**
     * Feature that draws resource time ranges, shaded areas displayed behind events. These zones are similar to events in
     * that they have a start and end date but different in that they do not take part in the event layout, and they always
     * occupy full row height.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/ResourceTimeRanges)
     */
    resourceTimeRangesFeature? : object|boolean|string|ResourceTimeRanges|ResourceTimeRangesConfig
    /**
     * Allow using [Ctrl/CMD + C/X] and [Ctrl/CMD + V] to copy/cut-and-paste rows. Also makes cut, copy and paste actions
     * available via the cell context menu.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/RowCopyPaste)
     */
    rowCopyPasteFeature? : object|boolean|string|RowCopyPaste|RowCopyPasteConfig
    /**
     * Enables expanding of Grid rows by either row click or double click, or by adding a separate Grid column which renders
     * a button that expands or collapses the row.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/RowExpander)
     */
    rowExpanderFeature? : object|boolean|string|RowExpander|RowExpanderConfig
    /**
     * This feature implements support for project transactions and used by default in Gantt. For general RowReorder feature
     * documentation see [RowReorder](https://bryntum.com/products/gantt/docs/api/Grid/feature/RowReorder).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/RowReorder)
     */
    rowReorderFeature? : object|boolean|string|RowReorder|RowReorderConfig
    /**
     * Enables user to change row height by dragging the bottom row border. After a resize operation, the [rowHeight](https://bryntum.com/products/gantt/docs/api/Grid/data/GridRowModel#field-rowHeight)
     * field of the record is updated (when [applyToAllRows](https://bryntum.com/products/gantt/docs/api/Grid/feature/RowResize#config-applyToAllRows) is `false`).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/RowResize)
     */
    rowResizeFeature? : object|boolean|string|GridRowResize|GridRowResizeConfig
    /**
     * Feature that displays a tooltip containing the time at the mouse position when hovering empty parts of the schedule.
     * To hide the schedule tooltip, just disable this feature:
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/ScheduleTooltip)
     */
    scheduleTooltipFeature? : object|boolean|string|ScheduleTooltip|ScheduleTooltipConfig
    /**
     * <div class="external-example " data-file="Grid/feature/Search.js"></div>
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/Search)
     */
    searchFeature? : object|boolean|string|Search|SearchConfig
    /**
     * Allows sorting of grid by clicking (or tapping) headers, also displays which columns grid is sorted by (numbered if
     * using multisort). Use modifier keys for multisorting: [Ctrl/CMD + click] to add sorter, [Ctrl/CMD + Alt + click] to remove sorter.
     * The actual sorting is done by the store, see [Store.sort()](https://bryntum.com/products/gantt/docs/api/Core/data/mixin/StoreSort#function-sort).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/Sort)
     */
    sortFeature? : object|boolean|string|Sort|SortConfig
    /**
     * This feature allows splitting the Scheduler into multiple views, either by using the cell context menu, or
     * programmatically by calling [split()](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/Split#function-split).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/Split)
     */
    splitFeature? : object|boolean|string|Split|SplitConfig
    /**
     * A feature which pins configurable content from a grid row to the top of the grid
     * while the row scrolls off the top but is still visible.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/StickyCells)
     */
    stickyCellsFeature? : object|boolean|string|StickyCells|StickyCellsConfig
    /**
     * Stripes rows by adding alternating CSS classes to all row elements (`b-even` and `b-odd`).
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/Stripe)
     */
    stripeFeature? : object|boolean|string|Stripe|StripeConfig
    /**
     * A special version of the Grid Summary feature. This feature displays a summary row in the grid footer.
     * For regular columns in the locked section - specify type of summary on columns, available types are:
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/Summary)
     */
    summaryFeature? : object|boolean|string|SchedulerSummary|SchedulerSummaryConfig
    /**
     * Adds scheduler specific menu items to the timeline header context menu.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/TimeAxisHeaderMenu)
     */
    timeAxisHeaderMenuFeature? : object|boolean|string|TimeAxisHeaderMenu|TimeAxisHeaderMenuConfig
    /**
     * Feature that renders global ranges of time in the timeline. Use this feature to visualize a `range` like a 1 hr lunch
     * or some important point in time (a `line`, i.e. a range with 0 duration). This feature can also show a current time
     * indicator if you set [showCurrentTimeLine](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/TimeRanges#config-showCurrentTimeLine) to true. To style the rendered elements, use the
     * [cls](https://bryntum.com/products/gantt/docs/api/Scheduler/model/TimeSpan#field-cls) field of the `TimeSpan` class.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Scheduler/feature/TimeRanges)
     */
    timeRangesFeature? : object|boolean|string|TimeRanges|TimeRangesConfig
    /**
     * This feature exposes methods on the owning Scheduler or Gantt widget which you can use to highlight one or multiple time spans
     * in the schedule. Please see [highlightTimeSpan](https://bryntum.com/products/gantt/docs/api/SchedulerPro/feature/TimeSpanHighlight#function-highlightTimeSpan) and [highlightTimeSpans](#SchedulerPro/feature/TimeSpanHighlight#function-highlightTimeSpans) to learn
     * more or try the demo below:
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/SchedulerPro/feature/TimeSpanHighlight)
     */
    timeSpanHighlightFeature? : object|boolean|string|TimeSpanHighlight|TimeSpanHighlightConfig
    /**
     * Feature that makes the grid work more like a tree. Included by default in [TreeGrid](https://bryntum.com/products/gantt/docs/api/Grid/view/TreeGrid). Requires
     * exactly one [TreeColumn](https://bryntum.com/products/gantt/docs/api/Grid/column/TreeColumn) among grids columns. That column will have its renderer replaced with a
     * tree renderer that adds padding and icon to give the appearance of a tree. The original renderer is preserved and
     * also called.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/Tree)
     */
    treeFeature? : object|boolean|string|Tree|TreeConfig
    /**
     * A feature that allows transforming a flat dataset (or the leaves of a hierarchical) into a tree by specifying a
     * record field per parent level. Parents are generated based on each leaf's value for those fields.
     * ...
     * [View online docs...](https://bryntum.com/products/gantt/docs/api/Grid/feature/TreeGroup)
     */
    treeGroupFeature? : object|boolean|string|GridTreeGroup|GridTreeGroupConfig

}

export class BryntumResourceUtilization extends React.Component<BryntumResourceUtilizationProps> {

    static instanceClass = ResourceUtilization;

    static instanceName = 'ResourceUtilization';

    processWidgetContent = processWidgetContent;

    // Component is View class
    static isView = true;

    // Hidden div to host unused portals DOM
    portalsCache? : HTMLElement;

    // CSS class used for individual portal containers
    portalContainerClass = 'b-react-portal-container';

    state = {
        // Holds React portals
        portals : new Map(),

        // Needed to trigger refresh when portals change
        generation : 0
    };

    static featureNames = [
        'cellCopyPasteFeature',
        'cellEditFeature',
        'cellMenuFeature',
        'cellTooltipFeature',
        'columnAutoWidthFeature',
        'columnDragToolbarFeature',
        'columnLinesFeature',
        'columnPickerFeature',
        'columnRenameFeature',
        'columnReorderFeature',
        'columnResizeFeature',
        'excelExporterFeature',
        'fileDropFeature',
        'fillHandleFeature',
        'filterFeature',
        'filterBarFeature',
        'groupFeature',
        'groupSummaryFeature',
        'headerMenuFeature',
        'headerZoomFeature',
        'lockRowsFeature',
        'mergeCellsFeature',
        'nonWorkingTimeFeature',
        'panFeature',
        'pdfExportFeature',
        'printFeature',
        'quickFindFeature',
        'regionResizeFeature',
        'resourceMenuFeature',
        'resourceNonWorkingTimeFeature',
        'resourceTimeRangesFeature',
        'rowCopyPasteFeature',
        'rowExpanderFeature',
        'rowReorderFeature',
        'rowResizeFeature',
        'scheduleTooltipFeature',
        'searchFeature',
        'sortFeature',
        'splitFeature',
        'stickyCellsFeature',
        'stripeFeature',
        'summaryFeature',
        'timeAxisHeaderMenuFeature',
        'timeRangesFeature',
        'timeSpanHighlightFeature',
        'treeFeature',
        'treeGroupFeature'
    ];

    projectStores = false;

    dataStores = {
        'originalStore': 'originals',
        'resourceTimeRangeStore': 'resourceTimeRanges',
        'timeRangeStore': 'timeRanges'
    };

    static configNames = [
        'adopt',
        'aggregateDataEntry',
        'aggregateHistogramDataForGroups',
        'animateRemovingRows',
        'animateTreeNodeToggle',
        'ariaDescription',
        'ariaLabel',
        'autoAdjustTimeAxis',
        'autoHeight',
        'barTextEffortUnit',
        'barTipEffortUnit',
        'barTooltipTemplate',
        'bbar',
        'bodyCls',
        'bubbleEvents',
        'bufferCoef',
        'bufferThreshold',
        'callOnFunctions',
        'collapsible',
        'config',
        'contentElementCls',
        'contextMenuTriggerEvent',
        'dataModelField',
        'defaultRegion',
        'defaultResourceImageName',
        'destroyStore',
        'destroyStores',
        'disableGridColumnIdWarning',
        'disableGridRowModelWarning',
        'dock',
        'effortFormat',
        'effortUnit',
        'enableSticky',
        'enableTextSelection',
        'endParamName',
        'fillLastColumn',
        'fixedRowHeight',
        'footer',
        'fullRowRefresh',
        'getBarClass',
        'getBarDOMConfig',
        'getBarText',
        'getBarTextDOMConfig',
        'getDataEntryForAggregating',
        'getOutlineClass',
        'getOutlineDOMConfig',
        'getRecordData',
        'getRowHeight',
        'hardRefreshOnTimeAxisReconfigure',
        'header',
        'hideHorizontalScrollbar',
        'histogramWidgetClass',
        'hoverCls',
        'ignoreDomEventsWhileScrolling',
        'ignoreParentReadOnly',
        'includeInactiveEvents',
        'infiniteScroll',
        'initAggregatedDataEntry',
        'listeners',
        'loadMask',
        'loadMaskDefaults',
        'loadMaskError',
        'localizable',
        'managedEventSizing',
        'maskDefaults',
        'masked',
        'maxTimeAxisUnit',
        'monitorResize',
        'owner',
        'partner',
        'passStartEndParameters',
        'plugins',
        'preserveFocusOnDatasetChange',
        'preserveScrollOnDatasetChange',
        'preventTooltipOnTouch',
        'relayStoreEvents',
        'resizeToFitIncludesHeader',
        'resourceImageExtension',
        'resourceImagePath',
        'resourceTimeRangeRenderer',
        'responsiveLevels',
        'ripple',
        'rootElement',
        'scaleColumn',
        'scalePoints',
        'scrollerClass',
        'scrollManager',
        'series',
        'showBarText',
        'showDirty',
        'showEffortUnit',
        'snapRelativeToEventStartDate',
        'startParamName',
        'stateful',
        'statefulEvents',
        'stateId',
        'stateProvider',
        'stickyHeaders',
        'strips',
        'subGridConfigs',
        'suppressFit',
        'syncMask',
        'tab',
        'tbar',
        'timeAxis',
        'type',
        'ui',
        'visibleZoomFactor',
        'weekStartDay',
        'weight',
        'zoomKeepsOriginalTimespan',
        'zoomOnMouseWheel',
        'zoomOnTimeAxisDoubleClick'
    ];

    static propertyConfigNames = [
        'alignSelf',
        'appendTo',
        'ariaLive',
        'barMargin',
        'cls',
        'collapsed',
        'columnLines',
        'columns',
        'dataset',
        'disabled',
        'displayDateFormat',
        'emptyText',
        'enableEventAnimations',
        'enableUndoRedoKeys',
        'endDate',
        'extraData',
        'fillTicks',
        'flex',
        'forceFit',
        'height',
        'hidden',
        'hideFooters',
        'hideHeaders',
        'hideRowHover',
        'histogramWidget',
        'id',
        'insertBefore',
        'insertFirst',
        'keyMap',
        'longPressTime',
        'margin',
        'maxHeight',
        'maxWidth',
        'maxZoomLevel',
        'minHeight',
        'minWidth',
        'minZoomLevel',
        'onBeforeAssignmentDelete',
        'onBeforeCancelCellEdit',
        'onBeforeCellDelete',
        'onBeforeCellEditStart',
        'onBeforeColumnDragStart',
        'onBeforeColumnDropFinalize',
        'onBeforeColumnResize',
        'onBeforeCopy',
        'onBeforeCSVExport',
        'onBeforeDestroy',
        'onBeforeEventDelete',
        'onBeforeExcelExport',
        'onBeforeFillHandleDragStart',
        'onBeforeFinishCellEdit',
        'onBeforeHide',
        'onBeforeHistogramDataCacheSet',
        'onBeforePan',
        'onBeforePaste',
        'onBeforePdfExport',
        'onBeforePresetChange',
        'onBeforeRenderHistogramRow',
        'onBeforeRenderRecordHistogram',
        'onBeforeRenderRow',
        'onBeforeRenderRows',
        'onBeforeRowCollapse',
        'onBeforeRowExpand',
        'onBeforeSelectionChange',
        'onBeforeSetRecord',
        'onBeforeShow',
        'onBeforeToggleGroup',
        'onBeforeToggleNode',
        'onCancelCellEdit',
        'onCatchAll',
        'onCellClick',
        'onCellContextMenu',
        'onCellDblClick',
        'onCellMenuBeforeShow',
        'onCellMenuItem',
        'onCellMenuShow',
        'onCellMenuToggleItem',
        'onCellMouseEnter',
        'onCellMouseLeave',
        'onCellMouseOut',
        'onCellMouseOver',
        'onCollapse',
        'onCollapseNode',
        'onColumnDrag',
        'onColumnDragStart',
        'onColumnDrop',
        'onColumnResize',
        'onColumnResizeStart',
        'onContextMenuItem',
        'onContextMenuToggleItem',
        'onCopy',
        'onCurrentTimelineUpdate',
        'onDataChange',
        'onDateRangeChange',
        'onDestroy',
        'onDirtyStateChange',
        'onDragSelecting',
        'onExpand',
        'onExpandNode',
        'onFileDrop',
        'onFillHandleBeforeDragFinalize',
        'onFillHandleDrag',
        'onFillHandleDragAbort',
        'onFillHandleDragEnd',
        'onFillHandleDragStart',
        'onFinishCellEdit',
        'onFocusIn',
        'onFocusOut',
        'onGridRowBeforeDragStart',
        'onGridRowBeforeDropFinalize',
        'onGridRowDrag',
        'onGridRowDragAbort',
        'onGridRowDragStart',
        'onGridRowDrop',
        'onHeaderClick',
        'onHeaderMenuBeforeShow',
        'onHeaderMenuItem',
        'onHeaderMenuShow',
        'onHeaderMenuToggleItem',
        'onHide',
        'onHistogramDataCacheSet',
        'onLockRows',
        'onMouseOut',
        'onMouseOver',
        'onPaint',
        'onPaste',
        'onPdfExport',
        'onPresetChange',
        'onReadOnly',
        'onRecompose',
        'onRenderRow',
        'onRenderRows',
        'onResize',
        'onResourceMenuBeforeShow',
        'onResourceMenuItem',
        'onResourceMenuShow',
        'onResourceMenuToggleItem',
        'onResourceNonWorkingTimeClick',
        'onResourceNonWorkingTimeContextMenu',
        'onResourceNonWorkingTimeDblClick',
        'onResourceNonWorkingTimeMouseDown',
        'onResourceNonWorkingTimeMouseOut',
        'onResourceNonWorkingTimeMouseOver',
        'onResourceNonWorkingTimeMouseUp',
        'onResourceTimeRangeClick',
        'onResourceTimeRangeContextMenu',
        'onResourceTimeRangeDblClick',
        'onResourceTimeRangeMouseDown',
        'onResourceTimeRangeMouseOut',
        'onResourceTimeRangeMouseOver',
        'onResourceTimeRangeMouseUp',
        'onResponsive',
        'onRowCollapse',
        'onRowExpand',
        'onRowMouseEnter',
        'onRowMouseLeave',
        'onScroll',
        'onSelectionChange',
        'onSelectionModeChange',
        'onShow',
        'onSplit',
        'onSplitterCollapseClick',
        'onSplitterDragEnd',
        'onSplitterDragStart',
        'onSplitterExpandClick',
        'onStartCellEdit',
        'onSubGridCollapse',
        'onSubGridExpand',
        'onTickSizeChange',
        'onTimeAxisChange',
        'onTimeAxisHeaderClick',
        'onTimeAxisHeaderContextMenu',
        'onTimeAxisHeaderDblClick',
        'onTimeAxisHeaderMenuBeforeShow',
        'onTimeAxisHeaderMenuItem',
        'onTimeAxisHeaderMenuShow',
        'onTimelineContextChange',
        'onTimelineViewportResize',
        'onTimeRangeHeaderClick',
        'onTimeRangeHeaderContextMenu',
        'onTimeRangeHeaderDblClick',
        'onToggleGroup',
        'onToggleNode',
        'onToolClick',
        'onUnlockRows',
        'onUnsplit',
        'onVisibleDateRangeChange',
        'preserveScroll',
        'presets',
        'project',
        'readOnly',
        'recurrenceConfirmationPopup',
        'resourceMargin',
        'resourceTimeRanges',
        'resourceTimeRangeStore',
        'rowHeight',
        'rowLines',
        'scrollable',
        'selectionMode',
        'showBarTip',
        'snap',
        'startDate',
        'store',
        'tickSize',
        'timeRanges',
        'timeRangeStore',
        'title',
        'tools',
        'transitionDuration',
        'viewPreset',
        'visibleDate',
        'width',
        'workingTime'
    ];

    static propertyNames = [
        'accessibility',
        'crudManager',
        'data',
        'eventColor',
        'eventStyle',
        'hasChanges',
        'isEngineReady',
        'originalStore',
        'parent',
        'scrollLeft',
        'scrollTop',
        'scrollX',
        'selectedCell',
        'selectedCells',
        'selectedRecord',
        'selectedRecords',
        'selectedRows',
        'state',
        'timeResolution',
        'timeZone',
        'tooltip',
        'zoomLevel'
    ];

    // Component instance
    instance!: ResourceUtilization;

    // Component element
    element! : HTMLElement;

    componentDidMount(): void {
        this.instance = createWidget(this);
    }

    componentWillUnmount(): void {
        // @ts-ignore
        this.instance?.destroy?.();
    }

    /**
     * Component about to be updated, from changing a prop using state.
     * React to it depending on what changed and prevent react from re-rendering our component.
     * @param nextProps
     * @param nextState
     * @returns {boolean}
     */
    shouldComponentUpdate(nextProps: Readonly<BryntumResourceUtilizationProps>, nextState: Readonly<{}>): boolean {
        return shouldComponentUpdate(this, nextProps, nextState);
    }

    render(): React.ReactNode {

        const className = `b-react-resource-utilization-container`;
        return (
            <React.Fragment>
                <div className={className} ref={(element) => (this.element = element!)}>
                    {Array.from(this.state.portals).map((item) => item[1])}
                </div>
                <div ref={el => (this.portalsCache = el!)} className="b-react-portals-cache" style={{ display : 'none' }}></div>
            </React.Fragment>
        );

    }
}
