// @bryntum/gantt-react/lib/BryntumGantt.js
import React from 'react';
import { Gantt } from '@bryntum/gantt';
import { createWidget, shouldComponentUpdate, processWidgetContent } from './WrapperHelper.js';
export class BryntumGantt extends React.Component {
    constructor() {
        super(...arguments);
        this.processWidgetContent = processWidgetContent;
        this.portalContainerClass = 'b-react-portal-container';
        this.state = {
            portals: new Map(),
            generation: 0
        };
        this.projectStores = true;
        this.dataStores = {
            'assignmentStore': 'assignments',
            'calendarManagerStore': 'calendars',
            'dependencyStore': 'dependencies',
            'eventStore': 'events',
            'resourceStore': 'resources',
            'taskStore': 'tasks',
            'timeRangeStore': 'timeRanges'
        };
    }
    componentDidMount() {
        this.instance = createWidget(this);
    }
    componentWillUnmount() {
        var _a, _b;
        (_b = (_a = this.instance) === null || _a === void 0 ? void 0 : _a.destroy) === null || _b === void 0 ? void 0 : _b.call(_a);
    }
    shouldComponentUpdate(nextProps, nextState) {
        return shouldComponentUpdate(this, nextProps, nextState);
    }
    render() {
        const className = `b-react-gantt-container`;
        return (React.createElement(React.Fragment, null,
            React.createElement("div", { className: className, ref: (element) => (this.element = element) }, Array.from(this.state.portals).map((item) => item[1])),
            React.createElement("div", { ref: el => (this.portalsCache = el), className: "b-react-portals-cache", style: { display: 'none' } })));
    }
}
BryntumGantt.instanceClass = Gantt;
BryntumGantt.instanceName = 'Gantt';
BryntumGantt.isView = true;
BryntumGantt.featureNames = [
    'baselinesFeature',
    'cellCopyPasteFeature',
    'cellEditFeature',
    'cellMenuFeature',
    'cellTooltipFeature',
    'columnAutoWidthFeature',
    'columnDragToolbarFeature',
    'columnLinesFeature',
    'columnPickerFeature',
    'columnRenameFeature',
    'columnReorderFeature',
    'columnResizeFeature',
    'criticalPathsFeature',
    'dependenciesFeature',
    'dependencyEditFeature',
    'eventFilterFeature',
    'eventSegmentsFeature',
    'excelExporterFeature',
    'fileDropFeature',
    'fillHandleFeature',
    'filterFeature',
    'filterBarFeature',
    'groupFeature',
    'groupSummaryFeature',
    'headerMenuFeature',
    'headerZoomFeature',
    'indicatorsFeature',
    'labelsFeature',
    'lockRowsFeature',
    'mergeCellsFeature',
    'mspExportFeature',
    'nonWorkingTimeFeature',
    'panFeature',
    'parentAreaFeature',
    'pdfExportFeature',
    'percentBarFeature',
    'printFeature',
    'progressLineFeature',
    'projectLinesFeature',
    'quickFindFeature',
    'regionResizeFeature',
    'rollupsFeature',
    'rowCopyPasteFeature',
    'rowExpanderFeature',
    'rowReorderFeature',
    'rowResizeFeature',
    'scheduleMenuFeature',
    'scheduleTooltipFeature',
    'scrollButtonsFeature',
    'searchFeature',
    'sortFeature',
    'splitFeature',
    'stickyCellsFeature',
    'stripeFeature',
    'summaryFeature',
    'taskCopyPasteFeature',
    'taskDragFeature',
    'taskDragCreateFeature',
    'taskEditFeature',
    'taskMenuFeature',
    'taskNonWorkingTimeFeature',
    'taskResizeFeature',
    'taskSegmentDragFeature',
    'taskSegmentResizeFeature',
    'taskTooltipFeature',
    'timeAxisHeaderMenuFeature',
    'timeRangesFeature',
    'timeSpanHighlightFeature',
    'treeFeature',
    'treeGroupFeature',
    'versionsFeature'
];
BryntumGantt.configNames = [
    'adopt',
    'allowCreate',
    'allowCreateOnlyParent',
    'animateRemovingRows',
    'animateTreeNodeToggle',
    'ariaDescription',
    'ariaLabel',
    'autoAdjustTimeAxis',
    'autoHeight',
    'bbar',
    'bodyCls',
    'bubbleEvents',
    'bufferCoef',
    'bufferThreshold',
    'callOnFunctions',
    'collapsible',
    'config',
    'contentElementCls',
    'contextMenuTriggerEvent',
    'creationTooltip',
    'cycleResolutionPopupClass',
    'defaultRegion',
    'defaultResourceImageName',
    'dependencyIdField',
    'destroyStore',
    'disableGridColumnIdWarning',
    'disableGridRowModelWarning',
    'displaySchedulingIssueResolutionPopup',
    'dock',
    'durationDisplayPrecision',
    'enableDeleteKey',
    'enableRecurringEvents',
    'enableSticky',
    'enableTextSelection',
    'enableTransactionalFeatures',
    'fillLastColumn',
    'fixedRowHeight',
    'footer',
    'fullRowRefresh',
    'getDateConstraints',
    'getRowHeight',
    'header',
    'hideHorizontalScrollbar',
    'hoverCls',
    'ignoreDomEventsWhileScrolling',
    'ignoreParentReadOnly',
    'infiniteScroll',
    'listeners',
    'loadMask',
    'loadMaskDefaults',
    'loadMaskError',
    'localizable',
    'managedEventSizing',
    'maskDefaults',
    'masked',
    'maxTimeAxisUnit',
    'monitorResize',
    'newTaskDefaults',
    'owner',
    'partner',
    'plugins',
    'preserveFocusOnDatasetChange',
    'preserveScrollOnDatasetChange',
    'preventTooltipOnTouch',
    'projectProgressReporting',
    'relayStoreEvents',
    'resizeToFitIncludesHeader',
    'resourceImageFolderPath',
    'responsiveLevels',
    'ripple',
    'rootElement',
    'schedulingIssueResolutionPopupClass',
    'scrollerClass',
    'scrollManager',
    'showCreationTooltip',
    'showDirty',
    'showLagInTooltip',
    'showTaskColorPickers',
    'snapRelativeToEventStartDate',
    'stateful',
    'statefulEvents',
    'stateId',
    'stateProvider',
    'stickyHeaders',
    'strips',
    'subGridConfigs',
    'suppressFit',
    'syncMask',
    'tab',
    'taskRenderer',
    'tbar',
    'terminalCls',
    'terminalSides',
    'timeAxis',
    'type',
    'ui',
    'visibleZoomFactor',
    'weekStartDay',
    'weight',
    'zoomKeepsOriginalTimespan',
    'zoomOnMouseWheel',
    'zoomOnTimeAxisDoubleClick'
];
BryntumGantt.propertyConfigNames = [
    'alignSelf',
    'allowDropOnEventBar',
    'appendTo',
    'ariaLive',
    'assignments',
    'barMargin',
    'calendars',
    'cls',
    'collapsed',
    'columnLines',
    'columns',
    'creationTooltipTemplate',
    'data',
    'dataset',
    'dependencies',
    'disabled',
    'displayDateFormat',
    'emptyText',
    'enableEventAnimations',
    'enableUndoRedoKeys',
    'endDate',
    'eventColor',
    'eventStyle',
    'extraData',
    'fillTicks',
    'flex',
    'forceFit',
    'height',
    'hidden',
    'hideFooters',
    'hideHeaders',
    'hideRowHover',
    'id',
    'insertBefore',
    'insertFirst',
    'keyMap',
    'longPressTime',
    'margin',
    'maxHeight',
    'maxWidth',
    'maxZoomLevel',
    'minHeight',
    'minWidth',
    'minZoomLevel',
    'onAfterDependencyCreateDrop',
    'onAfterDependencySave',
    'onAfterDragCreate',
    'onAfterEventSave',
    'onAfterTaskDrop',
    'onAfterTaskEdit',
    'onAfterTaskSave',
    'onAfterTaskSegmentDrop',
    'onBeforeAssignmentDelete',
    'onBeforeCancelCellEdit',
    'onBeforeCellDelete',
    'onBeforeCellEditStart',
    'onBeforeColumnDragStart',
    'onBeforeColumnDropFinalize',
    'onBeforeColumnResize',
    'onBeforeCopy',
    'onBeforeCSVExport',
    'onBeforeDependencyAdd',
    'onBeforeDependencyCreateDrag',
    'onBeforeDependencyCreateFinalize',
    'onBeforeDependencyDelete',
    'onBeforeDependencyEdit',
    'onBeforeDependencyEditShow',
    'onBeforeDependencySave',
    'onBeforeDestroy',
    'onBeforeDragCreate',
    'onBeforeDragCreateFinalize',
    'onBeforeEventDelete',
    'onBeforeEventResize',
    'onBeforeEventResizeFinalize',
    'onBeforeEventSave',
    'onBeforeEventSegmentResize',
    'onBeforeEventSegmentResizeFinalize',
    'onBeforeExcelExport',
    'onBeforeFillHandleDragStart',
    'onBeforeFinishCellEdit',
    'onBeforeHide',
    'onBeforeMspExport',
    'onBeforePan',
    'onBeforePaste',
    'onBeforePdfExport',
    'onBeforePresetChange',
    'onBeforeRenderRow',
    'onBeforeRenderRows',
    'onBeforeRowCollapse',
    'onBeforeRowExpand',
    'onBeforeSelectionChange',
    'onBeforeSetRecord',
    'onBeforeShow',
    'onBeforeShowTerminals',
    'onBeforeTaskAdd',
    'onBeforeTaskDelete',
    'onBeforeTaskDrag',
    'onBeforeTaskDropFinalize',
    'onBeforeTaskEdit',
    'onBeforeTaskEditShow',
    'onBeforeTaskResize',
    'onBeforeTaskResizeFinalize',
    'onBeforeTaskSave',
    'onBeforeTaskSegmentDrag',
    'onBeforeTaskSegmentDropFinalize',
    'onBeforeTaskSegmentResize',
    'onBeforeTaskSegmentResizeFinalize',
    'onBeforeToggleGroup',
    'onBeforeToggleNode',
    'onCancelCellEdit',
    'onCatchAll',
    'onCellClick',
    'onCellContextMenu',
    'onCellDblClick',
    'onCellMenuBeforeShow',
    'onCellMenuItem',
    'onCellMenuShow',
    'onCellMenuToggleItem',
    'onCellMouseEnter',
    'onCellMouseLeave',
    'onCellMouseOut',
    'onCellMouseOver',
    'onCollapse',
    'onCollapseNode',
    'onColumnDrag',
    'onColumnDragStart',
    'onColumnDrop',
    'onColumnResize',
    'onColumnResizeStart',
    'onContextMenuItem',
    'onContextMenuToggleItem',
    'onCopy',
    'onCriticalPathsHighlighted',
    'onCriticalPathsUnhighlighted',
    'onCurrentTimelineUpdate',
    'onDataChange',
    'onDateRangeChange',
    'onDependenciesDrawn',
    'onDependencyClick',
    'onDependencyContextMenu',
    'onDependencyCreateDragStart',
    'onDependencyCreateDrop',
    'onDependencyDblClick',
    'onDependencyMouseOut',
    'onDependencyMouseOver',
    'onDependencyValidationComplete',
    'onDependencyValidationStart',
    'onDestroy',
    'onDirtyStateChange',
    'onDragCreateEnd',
    'onDragCreateStart',
    'onDragSelecting',
    'onEventMenuBeforeShow',
    'onEventMenuItem',
    'onEventMenuShow',
    'onEventPartialResize',
    'onEventResizeEnd',
    'onEventResizeStart',
    'onEventSegmentPartialResize',
    'onEventSegmentResizeEnd',
    'onEventSegmentResizeStart',
    'onExpand',
    'onExpandNode',
    'onFileDrop',
    'onFillHandleBeforeDragFinalize',
    'onFillHandleDrag',
    'onFillHandleDragAbort',
    'onFillHandleDragEnd',
    'onFillHandleDragStart',
    'onFinishCellEdit',
    'onFocusIn',
    'onFocusOut',
    'onGridRowBeforeDragStart',
    'onGridRowBeforeDropFinalize',
    'onGridRowDrag',
    'onGridRowDragAbort',
    'onGridRowDragStart',
    'onGridRowDrop',
    'onHeaderClick',
    'onHeaderMenuBeforeShow',
    'onHeaderMenuItem',
    'onHeaderMenuShow',
    'onHeaderMenuToggleItem',
    'onHide',
    'onLockRows',
    'onMouseOut',
    'onMouseOver',
    'onMspExport',
    'onNavigate',
    'onPaint',
    'onPaste',
    'onPdfExport',
    'onPercentBarDrag',
    'onPercentBarDragAbort',
    'onPercentBarDragStart',
    'onPercentBarDrop',
    'onPresetChange',
    'onReadOnly',
    'onRecompose',
    'onReleaseTask',
    'onRenderRow',
    'onRenderRows',
    'onRenderTask',
    'onResize',
    'onResourceAssignmentClick',
    'onResponsive',
    'onRowCollapse',
    'onRowExpand',
    'onRowMouseEnter',
    'onRowMouseLeave',
    'onScheduleMenuBeforeShow',
    'onScheduleMenuItem',
    'onScheduleMenuShow',
    'onScroll',
    'onScrollButtonClick',
    'onSelectionChange',
    'onSelectionModeChange',
    'onShow',
    'onSplit',
    'onSplitterCollapseClick',
    'onSplitterDragEnd',
    'onSplitterDragStart',
    'onSplitterExpandClick',
    'onStartCellEdit',
    'onSubGridCollapse',
    'onSubGridExpand',
    'onTaskClick',
    'onTaskContextMenu',
    'onTaskDblClick',
    'onTaskDrag',
    'onTaskDragStart',
    'onTaskDrop',
    'onTaskEditCanceled',
    'onTaskKeyDown',
    'onTaskKeyUp',
    'onTaskMenuBeforeShow',
    'onTaskMenuItem',
    'onTaskMenuShow',
    'onTaskMouseDown',
    'onTaskMouseOut',
    'onTaskMouseOver',
    'onTaskMouseUp',
    'onTaskNonWorkingTimeClick',
    'onTaskNonWorkingTimeContextMenu',
    'onTaskNonWorkingTimeDblClick',
    'onTaskPartialResize',
    'onTaskResizeEnd',
    'onTaskResizeStart',
    'onTaskSegmentDrag',
    'onTaskSegmentDragStart',
    'onTaskSegmentDrop',
    'onTaskSegmentPartialResize',
    'onTaskSegmentResizeEnd',
    'onTaskSegmentResizeStart',
    'onTickSizeChange',
    'onTimeAxisChange',
    'onTimeAxisHeaderClick',
    'onTimeAxisHeaderContextMenu',
    'onTimeAxisHeaderDblClick',
    'onTimeAxisHeaderMenuBeforeShow',
    'onTimeAxisHeaderMenuItem',
    'onTimeAxisHeaderMenuShow',
    'onTimelineContextChange',
    'onTimelineViewportResize',
    'onTimeRangeHeaderClick',
    'onTimeRangeHeaderContextMenu',
    'onTimeRangeHeaderDblClick',
    'onToggleGroup',
    'onToggleNode',
    'onToolClick',
    'onTransactionChange',
    'onUnlockRows',
    'onUnsplit',
    'onVisibleDateRangeChange',
    'preserveScroll',
    'presets',
    'project',
    'readOnly',
    'recurrenceConfirmationPopup',
    'resources',
    'rowHeight',
    'rowLines',
    'scrollable',
    'scrollTaskIntoViewOnCellClick',
    'selectionMode',
    'showTooltip',
    'snap',
    'startDate',
    'store',
    'tasks',
    'taskStore',
    'terminalHideDelay',
    'terminalOffset',
    'terminalShowDelay',
    'terminalSize',
    'tickSize',
    'timeRanges',
    'timeZone',
    'title',
    'toggleParentTasksOnClick',
    'tools',
    'tooltipTemplate',
    'transitionDuration',
    'viewPreset',
    'visibleDate',
    'width',
    'workingTime'
];
BryntumGantt.propertyNames = [
    'accessibility',
    'hasChanges',
    'originalStore',
    'parent',
    'scrollLeft',
    'scrollTop',
    'scrollX',
    'selectedCell',
    'selectedCells',
    'selectedRecord',
    'selectedRecords',
    'selectedRows',
    'state',
    'timeResolution',
    'tooltip',
    'zoomLevel'
];
