// @bryntum/gantt-react/lib/BryntumTimelineHistogram.js
import React from 'react';
import { TimelineHistogram } from '@bryntum/gantt';
import { createWidget, shouldComponentUpdate, processWidgetContent } from './WrapperHelper.js';
export class BryntumTimelineHistogram extends React.Component {
    constructor() {
        super(...arguments);
        this.processWidgetContent = processWidgetContent;
        this.portalContainerClass = 'b-react-portal-container';
        this.state = {
            portals: new Map(),
            generation: 0
        };
        this.projectStores = false;
        this.dataStores = {
            'originalStore': 'originals'
        };
    }
    componentDidMount() {
        this.instance = createWidget(this);
    }
    componentWillUnmount() {
        var _a, _b;
        (_b = (_a = this.instance) === null || _a === void 0 ? void 0 : _a.destroy) === null || _b === void 0 ? void 0 : _b.call(_a);
    }
    shouldComponentUpdate(nextProps, nextState) {
        return shouldComponentUpdate(this, nextProps, nextState);
    }
    render() {
        const className = `b-react-timeline-histogram-container`;
        return (React.createElement(React.Fragment, null,
            React.createElement("div", { className: className, ref: (element) => (this.element = element) }, Array.from(this.state.portals).map((item) => item[1])),
            React.createElement("div", { ref: el => (this.portalsCache = el), className: "b-react-portals-cache", style: { display: 'none' } })));
    }
}
BryntumTimelineHistogram.instanceClass = TimelineHistogram;
BryntumTimelineHistogram.instanceName = 'TimelineHistogram';
BryntumTimelineHistogram.isView = true;
BryntumTimelineHistogram.featureNames = [
    'cellCopyPasteFeature',
    'cellEditFeature',
    'cellMenuFeature',
    'cellTooltipFeature',
    'columnAutoWidthFeature',
    'columnDragToolbarFeature',
    'columnLinesFeature',
    'columnPickerFeature',
    'columnRenameFeature',
    'columnReorderFeature',
    'columnResizeFeature',
    'excelExporterFeature',
    'fileDropFeature',
    'fillHandleFeature',
    'filterFeature',
    'filterBarFeature',
    'groupFeature',
    'groupSummaryFeature',
    'headerMenuFeature',
    'headerZoomFeature',
    'lockRowsFeature',
    'mergeCellsFeature',
    'nonWorkingTimeFeature',
    'panFeature',
    'pdfExportFeature',
    'printFeature',
    'quickFindFeature',
    'regionResizeFeature',
    'resourceMenuFeature',
    'resourceTimeRangesFeature',
    'rowCopyPasteFeature',
    'rowExpanderFeature',
    'rowReorderFeature',
    'rowResizeFeature',
    'scheduleTooltipFeature',
    'searchFeature',
    'sortFeature',
    'splitFeature',
    'stickyCellsFeature',
    'stripeFeature',
    'summaryFeature',
    'timeAxisHeaderMenuFeature',
    'timeRangesFeature',
    'treeFeature',
    'treeGroupFeature'
];
BryntumTimelineHistogram.configNames = [
    'adopt',
    'aggregateDataEntry',
    'aggregateHistogramDataForGroups',
    'animateRemovingRows',
    'animateTreeNodeToggle',
    'ariaDescription',
    'ariaLabel',
    'autoAdjustTimeAxis',
    'autoHeight',
    'barTooltipTemplate',
    'bbar',
    'bodyCls',
    'bubbleEvents',
    'bufferCoef',
    'bufferThreshold',
    'callOnFunctions',
    'collapsible',
    'config',
    'contentElementCls',
    'contextMenuTriggerEvent',
    'dataModelField',
    'defaultRegion',
    'destroyStore',
    'disableGridColumnIdWarning',
    'disableGridRowModelWarning',
    'dock',
    'durationDisplayPrecision',
    'enableRecurringEvents',
    'enableSticky',
    'enableTextSelection',
    'fillLastColumn',
    'fixedRowHeight',
    'footer',
    'fullRowRefresh',
    'getBarClass',
    'getBarDOMConfig',
    'getBarText',
    'getBarTextDOMConfig',
    'getDataEntryForAggregating',
    'getOutlineClass',
    'getOutlineDOMConfig',
    'getRecordData',
    'getRowHeight',
    'hardRefreshOnTimeAxisReconfigure',
    'header',
    'hideHorizontalScrollbar',
    'histogramWidgetClass',
    'hoverCls',
    'ignoreDomEventsWhileScrolling',
    'ignoreParentReadOnly',
    'infiniteScroll',
    'initAggregatedDataEntry',
    'listeners',
    'loadMask',
    'loadMaskDefaults',
    'loadMaskError',
    'localizable',
    'managedEventSizing',
    'maskDefaults',
    'masked',
    'maxTimeAxisUnit',
    'monitorResize',
    'owner',
    'partner',
    'plugins',
    'preserveFocusOnDatasetChange',
    'preserveScrollOnDatasetChange',
    'preventTooltipOnTouch',
    'relayStoreEvents',
    'resizeToFitIncludesHeader',
    'resourceTimeRangeRenderer',
    'responsiveLevels',
    'ripple',
    'rootElement',
    'scaleColumn',
    'scalePoints',
    'scrollerClass',
    'scrollManager',
    'series',
    'showDirty',
    'snapRelativeToEventStartDate',
    'stateful',
    'statefulEvents',
    'stateId',
    'stateProvider',
    'stickyHeaders',
    'strips',
    'subGridConfigs',
    'suppressFit',
    'syncMask',
    'tab',
    'tbar',
    'timeAxis',
    'type',
    'ui',
    'visibleZoomFactor',
    'weekStartDay',
    'weight',
    'zoomKeepsOriginalTimespan',
    'zoomOnMouseWheel',
    'zoomOnTimeAxisDoubleClick'
];
BryntumTimelineHistogram.propertyConfigNames = [
    'alignSelf',
    'appendTo',
    'ariaLive',
    'barMargin',
    'cls',
    'collapsed',
    'columnLines',
    'columns',
    'data',
    'dataset',
    'disabled',
    'displayDateFormat',
    'emptyText',
    'enableEventAnimations',
    'enableUndoRedoKeys',
    'endDate',
    'eventColor',
    'eventStyle',
    'extraData',
    'fillTicks',
    'flex',
    'forceFit',
    'height',
    'hidden',
    'hideFooters',
    'hideHeaders',
    'hideRowHover',
    'histogramWidget',
    'id',
    'insertBefore',
    'insertFirst',
    'keyMap',
    'longPressTime',
    'margin',
    'maxHeight',
    'maxWidth',
    'maxZoomLevel',
    'minHeight',
    'minWidth',
    'minZoomLevel',
    'onBeforeAssignmentDelete',
    'onBeforeCancelCellEdit',
    'onBeforeCellDelete',
    'onBeforeCellEditStart',
    'onBeforeColumnDragStart',
    'onBeforeColumnDropFinalize',
    'onBeforeColumnResize',
    'onBeforeCopy',
    'onBeforeCSVExport',
    'onBeforeDestroy',
    'onBeforeEventDelete',
    'onBeforeExcelExport',
    'onBeforeFillHandleDragStart',
    'onBeforeFinishCellEdit',
    'onBeforeHide',
    'onBeforeHistogramDataCacheSet',
    'onBeforePan',
    'onBeforePaste',
    'onBeforePdfExport',
    'onBeforePresetChange',
    'onBeforeRenderHistogramRow',
    'onBeforeRenderRecordHistogram',
    'onBeforeRenderRow',
    'onBeforeRenderRows',
    'onBeforeRowCollapse',
    'onBeforeRowExpand',
    'onBeforeSelectionChange',
    'onBeforeSetRecord',
    'onBeforeShow',
    'onBeforeToggleGroup',
    'onBeforeToggleNode',
    'onCancelCellEdit',
    'onCatchAll',
    'onCellClick',
    'onCellContextMenu',
    'onCellDblClick',
    'onCellMenuBeforeShow',
    'onCellMenuItem',
    'onCellMenuShow',
    'onCellMenuToggleItem',
    'onCellMouseEnter',
    'onCellMouseLeave',
    'onCellMouseOut',
    'onCellMouseOver',
    'onCollapse',
    'onCollapseNode',
    'onColumnDrag',
    'onColumnDragStart',
    'onColumnDrop',
    'onColumnResize',
    'onColumnResizeStart',
    'onContextMenuItem',
    'onContextMenuToggleItem',
    'onCopy',
    'onCurrentTimelineUpdate',
    'onDataChange',
    'onDateRangeChange',
    'onDestroy',
    'onDirtyStateChange',
    'onDragSelecting',
    'onExpand',
    'onExpandNode',
    'onFileDrop',
    'onFillHandleBeforeDragFinalize',
    'onFillHandleDrag',
    'onFillHandleDragAbort',
    'onFillHandleDragEnd',
    'onFillHandleDragStart',
    'onFinishCellEdit',
    'onFocusIn',
    'onFocusOut',
    'onGridRowBeforeDragStart',
    'onGridRowBeforeDropFinalize',
    'onGridRowDrag',
    'onGridRowDragAbort',
    'onGridRowDragStart',
    'onGridRowDrop',
    'onHeaderClick',
    'onHeaderMenuBeforeShow',
    'onHeaderMenuItem',
    'onHeaderMenuShow',
    'onHeaderMenuToggleItem',
    'onHide',
    'onHistogramDataCacheSet',
    'onLockRows',
    'onMouseOut',
    'onMouseOver',
    'onPaint',
    'onPaste',
    'onPdfExport',
    'onPresetChange',
    'onReadOnly',
    'onRecompose',
    'onRenderRow',
    'onRenderRows',
    'onResize',
    'onResourceMenuBeforeShow',
    'onResourceMenuItem',
    'onResourceMenuShow',
    'onResourceMenuToggleItem',
    'onResourceTimeRangeClick',
    'onResourceTimeRangeContextMenu',
    'onResourceTimeRangeDblClick',
    'onResourceTimeRangeMouseDown',
    'onResourceTimeRangeMouseOut',
    'onResourceTimeRangeMouseOver',
    'onResourceTimeRangeMouseUp',
    'onResponsive',
    'onRowCollapse',
    'onRowExpand',
    'onRowMouseEnter',
    'onRowMouseLeave',
    'onScroll',
    'onSelectionChange',
    'onSelectionModeChange',
    'onShow',
    'onSplit',
    'onSplitterCollapseClick',
    'onSplitterDragEnd',
    'onSplitterDragStart',
    'onSplitterExpandClick',
    'onStartCellEdit',
    'onSubGridCollapse',
    'onSubGridExpand',
    'onTickSizeChange',
    'onTimeAxisChange',
    'onTimeAxisHeaderClick',
    'onTimeAxisHeaderContextMenu',
    'onTimeAxisHeaderDblClick',
    'onTimeAxisHeaderMenuBeforeShow',
    'onTimeAxisHeaderMenuItem',
    'onTimeAxisHeaderMenuShow',
    'onTimelineContextChange',
    'onTimelineViewportResize',
    'onTimeRangeHeaderClick',
    'onTimeRangeHeaderContextMenu',
    'onTimeRangeHeaderDblClick',
    'onToggleGroup',
    'onToggleNode',
    'onToolClick',
    'onUnlockRows',
    'onUnsplit',
    'onVisibleDateRangeChange',
    'preserveScroll',
    'presets',
    'readOnly',
    'recurrenceConfirmationPopup',
    'rowHeight',
    'rowLines',
    'scrollable',
    'selectionMode',
    'showBarTip',
    'snap',
    'startDate',
    'store',
    'tickSize',
    'timeZone',
    'title',
    'tools',
    'transitionDuration',
    'viewPreset',
    'visibleDate',
    'width',
    'workingTime'
];
BryntumTimelineHistogram.propertyNames = [
    'accessibility',
    'hasChanges',
    'originalStore',
    'parent',
    'scrollLeft',
    'scrollTop',
    'scrollX',
    'selectedCell',
    'selectedCells',
    'selectedRecord',
    'selectedRecords',
    'selectedRows',
    'state',
    'timeResolution',
    'tooltip',
    'zoomLevel'
];
